---
title: "<br>Workflow part II:<br>ASV Inference and Taxon Assignment"
output:
  rmarkdown::html_document:
    code_download: true
    theme: cerulean
    toc: true
    toc_float: true
    css: my.css
---

```{r knitr_options, echo=FALSE}
knitr::opts_chunk$set(results='hide',
                      eval = FALSE,
                      root.dir = "..",
                      class.source="r_code",
                      class.output="r_output",
                      class.warning="r_warning",
                      class.message="r_warning",
                      class.error="r_error")
```

<br>

-----

## Goals

Process metabarcoding reads, post-adapter removal, including:

- Evaluation of quality
- Quality trimming and filtering
- Error correction and denoising
- Read merging
- Assigning taxonomy

-----

## Before We Get Started

### Notes

- This documented was adapted from [Callahan et al. 2006](https://f1000research.com/articles/5-1492/v2)
  by Matthew Willman, with further edits by Soledad Benitez Ponce and Jelmer Poelstra."
  
- To convert an `Rmd` (R Markdown) file to an R script,
  type the following in an R console:
  `knitr::purl(input="<filename>.Rmd")`.
  (You can download this `Rmd` file by clicking the `Code` button
  in the top-right of this page -- but we will open it at OSC.)

### Start an RStudio Server job at OSC

- Login to OSC at <https://ondemand.osc.edu>.

- Click on `Interactive Apps` (top bar) > `RStudio Server (Owens and Pitzer)`

<p align="center">
<img src=slides/figs_OSC/apps.png width="500">
</p>

- Fill out the form as shown [here](slides/03-OSC-slides.html#rstudio_server_job).

- Once your job has started, click `Launch RStudio Server`.

- You should automatically be in your personal dir inside the dir
  '/fs/project/PAS0471/workshops/2020-12_micro'.
  (In the `Files` pane, you should see your `.Rproj` file. 
  If not, type `setwd('/fs/project/PAS0471/workshops/2020-12_micro')`
  and then click on the dir with your name in the `Files` pane).

- Now, click on the `markdown` directory in the Files pane,
  and open the file `06-reads-to-ASV.Rmd`. That's this file!

-----

## Step 1: Getting Started

### Install and load packages

To save time, we have already installed all the necessary R packages at OSC
into a custom library.
To add this library for the current R session:

```{r libload}
.libPaths(new = '/fs/project/PAS0471/.R/4.0/')
```


```{r}
packages <- c("ggplot2", "gridExtra",
              "dada2", "phyloseq", "DECIPHER", "phangorn")
pacman::p_load(char = packages, install = TRUE)
```

### Set the file paths

```{r}
# Dir with input fastq files:
indir <- "data/processed/fastq_trimmed/second_trim/"

# Dir for output:
outdir <- 'analysis/ASV_inference/'
dir.create(outdir, recursive = TRUE)

# Fasta file with training data:
# (Check for an up-to-date version at <https://benjjneb.github.io/dada2/training.html>)
tax_key <- "ref_data/silva_nr99_v138_train_set.fa" 

# File with sample metadata:
sample_names_file <- "metadata/sample_meta.txt"
```
