---
title: "<br>Workflow part III:<br>Analyzing ASV Data"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
    css: my.css
---

```{r knitr_options, echo=FALSE}
knitr::opts_chunk$set(results='hide',
                      eval = FALSE,
                      class.source="r_code",
                      class.output="r_output",
                      class.warning="r_warning",
                      class.message="r_warning",
                      class.error="r_error")
```

<br>

-----

## Goals

Analyze microbiome experimental data as a phyloseq object - explore ecological metrics and identify differentially abundant taxa.

-----

## Before We Get Started

### Notes

- This is a slightly modified version of a document created by [Matthew Willman](mailto:willman.18@osu.edu).

- To convert an `Rmd` (R Markdown) file to an R script,
  type the following in an R console:
  `knitr::purl(input="<filename>.Rmd")`.
  (You can download this `Rmd` file by clicking the `Code` button
  in the top-right of this page -- but we will open it at OSC.)

### Open this file in RStudio Server at OSC

- This assumes you still have an active RStudio Server job at OSC;
   if not see the
   [instructions on the previous page](07-ASV-inference.html#start-an-rstudio-server-job-at-osc) to start one.

- In the Files pane, in the `markdown` directory,
  open the file `08-postASV-analysis.Rmd`. That's this file!

 
-----

## Getting Started

### Install and Load Packages

To save time, we have already installed all the necessary R packages at OSC
into a custom library.
To add this library for the current R session:

```{r libload}
.libPaths(new = '/fs/project/PAS0471/.R/4.0/')
```

```{r packages, message=FALSE, warning=FALSE}
packages <- c('ggplot2', 'vegan', 'phyloseq', 'decontam', 'ape',
              'DESeq2', 'microbiome', 'metagenomeSeq', 'remotes',
              'ampvis2', 'breakaway')
pacman::p_load(char = packages, install = TRUE)

# If you wanted to install and load these packages yourself,
# first make sure you have the pacman package installed:
## install.packages('pacman')
# Then, the code above would work except for the last two packages,
# which are available from Github only. To install these, you would run:
## remotes::install_github("MadsAlbertsen/ampvis2")
## remotes::install_github("adw96/breakaway")
```

### Set Directories

```{r}
# Dir with input fastq files:
indir <- 'analysis/ASV_inference/master/'

# Dir for output:
outdir <- 'analysis/postASV_analysis/'
```

