<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title> Workflow part II: ASV Inference and Taxon Assignment</title>

<script src="site_libs/header-attrs-2.5/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
#rmd-source-code {
  display: none;
}
</style>

<link rel="stylesheet" href="my.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Microbiomics Workshop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Overview
  </a>
</li>
<li>
  <a href="00-setup.html">
    <span class="fa fa-gear"></span>
     
    Setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Lectures:</li>
    <li>
      <a href="01-microbiomics.html">Intro to Microbiomics</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Lecture/Code-along:</li>
    <li>
      <a href="02-shell.html">Intro to the Shell</a>
    </li>
    <li>
      <a href="03-OSC.html">Intro to OSC</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Workflow Sessions:</li>
    <li>
      <a href="04-preprocess-reads.html">I: Adapter and Primer Removal</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Lecture/Code-along:</li>
    <li>
      <a href="05-R.html">Intro to R</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Workflow Sessions:</li>
    <li>
      <a href="06-reads-to-ASV.html">II: ASV Infererence and Taxon Assignment</a>
    </li>
    <li>
      <a href="07-ASV-analysis.html">III: Analyzing ASV Data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/mcic-osu/2020-12-microbiomics-workshop.git">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore"><br>Workflow part II:<br>ASV Inference and Taxon Assignment</h1>

</div>


<p><br></p>
<hr />
<div id="goals" class="section level2">
<h2>Goals</h2>
<p>Process metabarcoding reads, post-adapter removal, including:</p>
<ul>
<li>Evaluation of quality</li>
<li>Quality trimming and filtering</li>
<li>Error correction and denoising</li>
<li>Read merging</li>
<li>Assigning taxonomy</li>
</ul>
<hr />
</div>
<div id="before-we-get-started" class="section level2">
<h2>Before We Get Started</h2>
<div id="notes" class="section level3">
<h3>Notes</h3>
<ul>
<li><p>This documented was adapted from <a href="https://f1000research.com/articles/5-1492/v2">Callahan et al. 2006</a> by Matthew Willman, with further edits by Soledad Benitez Ponce and Jelmer Poelstra."</p></li>
<li><p>To convert an <code>Rmd</code> (R Markdown) file to an R script, type the following in an R console: <code>knitr::purl(input="&lt;filename&gt;.Rmd")</code>. (You can download this <code>Rmd</code> file by clicking the <code>Code</code> button in the top-right of this page – but we will open it at OSC.)</p></li>
</ul>
</div>
<div id="start-an-rstudio-server-job-at-osc" class="section level3">
<h3>Start an RStudio Server job at OSC</h3>
<ul>
<li><p>Login to OSC at <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>.</p></li>
<li><p>Click on <code>Interactive Apps</code> (top bar) &gt; <code>RStudio Server (Owens and Pitzer)</code></p></li>
</ul>
<p align="center">
<img src=slides/figs_OSC/apps.png width="500">
</p>
<ul>
<li><p>Fill out the form as shown <a href="slides/03-OSC.html#rstudio_server_job">here</a></p></li>
<li><p>Once your job has started, click <code>Launch RStudio Server</code>.</p></li>
<li><p>In the R console, enter:</p>
<pre class="sh"><code>setwd(&#39;/fs/project/PAS0471/workshops/2020-12_micro&#39;)</code></pre></li>
<li><p>Then in the Files pane, in the bottom right, click the Gear icon and then <code>Go To Working Directory</code>, as shown below:</p></li>
</ul>
<p align="center">
<img src=img/RStudio_go-to-wd.png width="500">
</p>
<ul>
<li>Now, click on the <code>markdown</code> directory in the Files pane, and open the file <code>06-reads-to-ASV.Rmd</code>. That’s this file!</li>
</ul>
<hr />
</div>
</div>
<div id="step-1-getting-started" class="section level2">
<h2>Step 1: Getting Started</h2>
<div id="install-and-load-packages" class="section level3">
<h3>Install and load packages</h3>
<p>To save time, we have already installed all the necessary R packages at OSC into a custom library. To add this library for the current R session:</p>
<pre class="r r_code"><code>.libPaths(new = &#39;/fs/project/PAS0471/.R/4.0/&#39;)</code></pre>
<pre class="r r_code"><code>packages &lt;- c(&quot;ggplot2&quot;, &quot;gridExtra&quot;,
              &quot;dada2&quot;, &quot;phyloseq&quot;, &quot;DECIPHER&quot;, &quot;phangorn&quot;)
pacman::p_load(char = packages, install = TRUE)</code></pre>
</div>
<div id="set-the-file-paths" class="section level3">
<h3>Set the file paths</h3>
<pre class="r r_code"><code># Dir with input fastq files:
indir &lt;- &quot;processed_data/fastq_trimmed/second_trim/&quot;

# Dir for output:
outdir_base &lt;- &#39;analysis/ASV_inference/output/&#39;
me &lt;- system(&quot;echo $USER&quot;, intern = TRUE)
outdir &lt;- file.path(outdir_base, me)
dir.create(outdir, recursive = TRUE)
outdir

# Fasta file with training data:
# (Check for an up-to-date version at &lt;https://benjjneb.github.io/dada2/training.html&gt;)
tax_key &lt;- &quot;analysis/ASV_inference/input/silva_nr99_v138_train_set.fa&quot; 

# File with sample metadata:
sample_names_file &lt;- &quot;metadata/sample_meta.txt&quot;</code></pre>
</div>
<div id="assign-files-to-forward-and-reverse-reads" class="section level3">
<h3>Assign files to forward and reverse reads</h3>
<p>For forward and reverse fastq filenames which have the following name format: <code>&lt;SAMPLENAME&gt;_R1_001.fastq</code> and <code>&lt;SAMPLENAME&gt;_R2_001.fastq</code>.</p>
<pre class="r r_code"><code>fnFs &lt;- sort(list.files(indir, pattern=&quot;_R1_001.fastq.gz&quot;, full.names = TRUE))
fnRs &lt;- sort(list.files(indir, pattern=&quot;_R2_001.fastq.gz&quot;, full.names = TRUE))</code></pre>
</div>
<div id="extract-sample-names" class="section level3">
<h3>Extract sample names</h3>
<p>Extract the sample names from the <code>.fastq</code> file names, assuming file names have format: <code>&lt;SAMPLENAME&gt;_XXX.fastq</code>.</p>
<p>The resulting <code>sample.names</code> should match the SampleID column in the <code>sample_data_example</code> file.</p>
<pre class="r r_code"><code>sample.names &lt;- sapply(strsplit(basename(fnFs), &quot;-V4-V5_&quot;), `[`, 1)</code></pre>
<hr />
</div>
</div>
<div id="step-2-qc" class="section level2">
<h2>Step 2: QC</h2>
<div id="plot-sequence-quality-data" class="section level3">
<h3>Plot sequence quality data</h3>
<p>You can generate and evaluate plots individually:</p>
<pre class="r r_code"><code>plotQualityProfile(fnFs[1:2])</code></pre>
<p><br></p>
<p>Or generate pdf files to inspect multiple plots at once:</p>
<pre class="r r_code"><code>ii &lt;- sample(length(fnFs), 3)
pdf(&quot;Error_profiles.pdf&quot;)
par(mfrow=c(2,3))
for (i in ii) { print (plotQualityProfile(fnFs[i]) + ggtitle(&quot;Fwd&quot;)) }
ii &lt;- sample(length(fnRs), 3)
for (i in ii) { print (plotQualityProfile(fnRs[i]) + ggtitle(&quot;Rev&quot;)) }
dev.off()</code></pre>
<hr />
</div>
</div>
<div id="step-3-trimming" class="section level2">
<h2>Step 3: Trimming</h2>
<p>The <code>truncLen</code> parameters should be based on sequence quality visualized above. It is suggested to trim first 10 nucleotides of each read, empirical observations across Illumina datasets show these positions are likely to contain pathological errors. Trimming length could be different for R1 and R2, but ideally not too drastically so, in order to maximize merging pairs in subsequent steps.</p>
<p>First, define paths for trimming:</p>
<pre class="r r_code"><code>filtFs &lt;- file.path(output, &quot;filtered&quot;, paste0(sample.names, &quot;_F_filt.fastq&quot;))
filtRs &lt;- file.path(output, &quot;filtered&quot;, paste0(sample.names, &quot;_R_filt.fastq&quot;))</code></pre>
<p><br></p>
<p>Then, perform the trimming:</p>
<pre class="r r_code"><code>print(&quot;filter and Trimming&quot;)
Sys.time()
out &lt;- filterAndTrim(fnFs, filtFs, fnRs, filtRs, trimLeft=10, truncLen=c(250,210), maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=F, compress=FALSE, multithread=TRUE, verbose=TRUE) 
print(&quot;done&quot;)
Sys.time()</code></pre>
<hr />
</div>
<div id="step-4-dereplication-error-training" class="section level2">
<h2>Step 4: Dereplication &amp; Error Training</h2>
<pre class="r r_code"><code>derepFs &lt;- derepFastq(filtFs, verbose = FALSE)
derepRs &lt;- derepFastq(filtRs, verbose = FALSE)

print(&quot;Learning errors:&quot;)
Sys.time()
errF &lt;- learnErrors(derepFs, multithread = TRUE, verbose = TRUE)
errR &lt;- learnErrors(derepRs, multithread = TRUE, verbose = TRUE)
print(&quot;Done.&quot;)
Sys.time()

names(derepFs) &lt;- sample.names
names(derepRs) &lt;- sample.names</code></pre>
<p><br></p>
<p>Plot errors to verify that error rates have been reasonable well-estimated. Observe the fit between observed error rates (points) and fitted error rates (lines):</p>
<pre class="r r_code"><code>plotErrors(errF, nominalQ = TRUE)
plotErrors(errR, nominalQ = TRUE)</code></pre>
<hr />
</div>
<div id="step-5-infer-asvs" class="section level2">
<h2>Step 5: Infer ASVs</h2>
<p>We will run the dada algorithm to infer ASVs.</p>
<p>This step is more computationally intensity. Indepedant inference by sample (pool=F) will keep memory requirements linear to sample number. Pooling will increase computation time but improve detection of rare variants seen once or twice in an individual sample but many times across all samples.</p>
<pre class="r r_code"><code>print(&quot;inferring ASVs (running dada algorithm)&quot;)
Sys.time()
dadaFs &lt;- dada(derepFs, err = errF, multithread = T, pool = FALSE)
dadaRs &lt;- dada(derepRs, err = errR, multithread = T, pool = FALSE)
print(&quot;done&quot;)
Sys.time()

#inspect object
dadaFs[[1]]</code></pre>
<hr />
</div>
<div id="step-6-merge-read-pairs" class="section level2">
<h2>Step 6: Merge Read Pairs</h2>
<p>In this step, we will merge read pairs and generate sequence table of paired ASVs.</p>
<pre class="r r_code"><code>mergers &lt;- mergePairs(dadaFs, derepFs, dadaRs, derepRs,
                      verbose=TRUE)

seqtab.all &lt;- makeSequenceTable(mergers)
dim(seqtab.all)</code></pre>
<hr />
</div>
<div id="step-7-summarize-sequences-and-remove-chimeras" class="section level2">
<h2>Step 7: Summarize Sequences and Remove Chimeras</h2>
<pre class="r r_code"><code># Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab.all)))

#If you need to remove sequences of a particular length (e.g. too long):
#seqtab2 &lt;- seqtab[,nchar(colnames(seqtab.all)) %in% seq(250,256)]

# Remove chimeras:
seqtab &lt;- removeBimeraDenovo(seqtab.all,
                             method = &quot;consensus&quot;,
                             multithread = TRUE,
                             verbose = TRUE)
dim(seqtab)

# Proportion of removed sequences
sum(seqtab) / sum(seqtab.all)

# Save seqtab file as an RDS object
saveRDS(seqtab, file = file.path(outdir, &quot;seqtab_V4.rds&quot;))</code></pre>
<hr />
</div>
<div id="step-8-generate-a-summary-table" class="section level2">
<h2>Step 8: Generate a Summary Table</h2>
<p>In this step, we will generate a summary table of the number of sequences processed and outputs of different steps of the pipeline.</p>
<p>This information is generally used to further evaluate characteristics and quality of the run, sample-to-sample variation, and resulting sequencing depth for each sample.</p>
<pre class="r r_code"><code>getN &lt;- function(x) sum(getUniques(x))

track &lt;- cbind(out,
               sapply(dadaFs, getN),
               sapply(dadaRs, getN),
               sapply(mergers, getN),
               rowSums(seqtab))

# If processing a single sample, remove the sapply calls:
# e.g. replace sapply(dadaFs, getN) with getN(dadaFs)

colnames(track) &lt;- c(&quot;input&quot;, &quot;filtered&quot;, &quot;denoisedF&quot;,
                     &quot;denoisedR&quot;, &quot;merged&quot;, &quot;nonchim&quot;)
rownames(track) &lt;- sample.names
head(track)

outfile_track &lt;- file.path(outdir, &#39;track.txt&#39;)
write.table(track, file = outfile_track)</code></pre>
<hr />
</div>
<div id="step-9-assign-taxonomy-to-asvs" class="section level2">
<h2>Step 9: Assign Taxonomy to ASVs</h2>
<p>Depending on the marker gene and the data, you will choose the appropriate file for this step. Several files have been formatted for taxony assignments in DADA2 pipeline and are available in the following website: <a href="https://benjjneb.github.io/dada2/index.html" class="uri">https://benjjneb.github.io/dada2/index.html</a></p>
<pre class="r r_code"><code>print(&quot;Assigning taxa to ASVs&quot;)
Sys.time()
taxa &lt;- assignTaxonomy(seqtab, tax_key, multithread = TRUE)
colnames(taxa) &lt;- c(&quot;Kingdom&quot;, &quot;Phylum&quot;, &quot;Class&quot;, &quot;Order&quot;, &quot;Family&quot;, &quot;Genus&quot;)
print(&quot;done&quot;)
Sys.time()</code></pre>
<hr />
</div>
<div id="step-10-phylogenetic-tree-estimation" class="section level2">
<h2>Step 10: Phylogenetic Tree Estimation</h2>
<p>A phylogenetic tree can be estimated for the sequence data you generated. Depending on the number of ASVs recovered and the phylogenetic tree algorithm of choice, this step could take several days. Simpler trees will be less computationally intensive. Depending on the marker gene you are working on, you might choose or not to use this option.</p>
<pre class="r r_code"><code>#seqtab&lt;- readRDS(&quot;seqtab_V4.rds&quot;)
seqs &lt;- getSequences(seqtab)
names(seqs) &lt;- seqs # This propagates to the tip labels of the tree. At this stage ASV labels are full ASV sequence
alignment &lt;- AlignSeqs(DNAStringSet(seqs), anchor=NA, iterations = 5, refinements = 5)

print(&quot;Computing pairwise distances from ASVs&quot;)
Sys.time()
phang.align &lt;- phyDat(as(alignment, &quot;matrix&quot;), type=&quot;DNA&quot;)
dm &lt;- dist.ml(phang.align)
treeNJ &lt;- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phang.align)
print(&quot;done&quot;)
Sys.time()

fitGTR &lt;- update(fit, k=4, inv=0.2)
print(&quot;done&quot;)
Sys.time()

print(&quot;Computing likelihood of tree&quot;)
Sys.time()
fitGTR &lt;- optim.pml(fitGTR, model=&quot;GTR&quot;, optInv=TRUE, optGamma=TRUE,
                      rearrangement = &quot;stochastic&quot;, control = pml.control(trace = 0))
print(&quot;done&quot;)
Sys.time()</code></pre>
<hr />
</div>
<div id="step-11-generate-output-files" class="section level2">
<h2>Step 11: Generate Output Files</h2>
<p>In this step, we will generate output files formatted for downstream analysis in phyloseq</p>
<p>Data will be extracted from the DADA2 outputs and formatted for downstream analysis and data inspection. The last step of this pipeline saves a phyloseq object as an .RDS file and can be imported directly in phyloseq.</p>
<pre class="r r_code"><code># Rename ASV names: 
asv_seqs&lt;- colnames(seqtab)
asv_headers&lt;- vector(dim(seqtab)[2], mode = &quot;character&quot;)

for (i in 1:dim(seqtab)[2]) {
    asv_headers[i] &lt;- paste(&quot;&gt;ASV&quot;, i, sep = &quot;_&quot;)
}


# Write ASV fasta for final ASV seqs.
#This fasta file can also be used for phylogenetic tree inference in different platforms
asv_fasta&lt;- c(rbind(asv_headers, asv_seqs))

outfile_fasta &lt;- file.path(outdir, &#39;ASVs.fa&#39;)
write(asv_fasta, outfile_fasta)


# Count table:
asv_tab &lt;- t(seqtab)
row.names(asv_tab) &lt;- sub(&quot;&gt;&quot;, &quot;&quot;, asv_headers)

outfile_tab &lt;- file.path(outdir, &#39;ASVs_counts.tsv&#39;)
write.table(asv_tab, outfile_tab, 
            sep = &quot;\t&quot;, quote = FALSE, col.names = NA)


# Tax table:
asv_tax &lt;- taxa
row.names(asv_tax) &lt;- sub(&quot;&gt;&quot;, &quot;&quot;, asv_headers)

outfile_tax &lt;- file.path(outdir, &#39;ASVs_taxonomy.tsv&#39;)
write.table(asv_tax, outfile_tax,
            sep = &quot;\t&quot;, quote = FALSE, col.names = NA)


# Read files back in:
count_tab &lt;- read.table(outfile_tab,
                        header = TRUE, row.names = 1,
                        check.names = FALSE, sep = &quot;\t&quot;)

tax_tab &lt;- as.matrix(
  read.table(outfile_tax,
             header = TRUE, row.names = 1,
             check.names = FALSE, sep = &quot;\t&quot;)
  )


# Load and prepare sample metadata:
sample_df &lt;- read.table(file = sample_names_file,
                        sep = &quot;\t&quot;, header = TRUE)
colnames(samdf)[1]&lt;- &quot;SampleID&quot;
rownames(samdf)&lt;- samdf$SampleID


# Build the final phyloseq object:
# We will not do this now, but a phylogenetic tree can also be added to the phyloseq object.
ps &lt;- phyloseq(otu_table(count_tab, taxa_are_rows = TRUE),
               sample_data(sample_df),
               tax_table(tax_tab))

outfile_ps &lt;- file.path(outdir, &#39;ps_V4.rds&#39;)
saveRDS(ps, file = outfile_ps)

print(&quot;done&quot;)</code></pre>
<hr />
</div>
<div id="resources" class="section level2">
<h2>Resources</h2>
<ul>
<li><a href="https://f1000research.com/articles/5-1492/v2">Callahan et al. 2006: “Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses”</a></li>
<li><a href="https://benjjneb.github.io/dada2/index.html"><code>dada2</code> documentation and tutorials</a></li>
<li><a href="https://benjjneb.github.io/dada2/training.html">Taxonomic references for <code>dada2</code></a></li>
<li><a href="https://cutadapt.readthedocs.io/en/stable/index.html"><code>cutadapt</code> documentation and tutorials</a></li>
</ul>
<hr />
</div>
<div id="bonus-submit-script-as-an-osc-job" class="section level2">
<h2>Bonus: Submit script as an OSC job</h2>
<ul>
<li>Script <code>dada2.sh</code>:</li>
</ul>
<pre class="sh"><code>#PBS -l walltime=100:00:00
#PBS -l nodes=1:ppn=28
#PBS -A PAS1548

module load gnu/9.1.0
module load mkl/2019.0.3
module load R/3.6.1
cd /fs/scratch/PAS1548/BaseSpace_out/11-21-19/V4_V5/scripts
Rscript dada2_V4.R</code></pre>
</div>

<div id="rmd-source-code">---
title: "<br>Workflow part II:<br>ASV Inference and Taxon Assignment"
output:
  rmarkdown::html_document:
    code_download: true
    theme: cerulean
    toc: true
    toc_float: true
    css: my.css
---

```{r knitr_options, echo=FALSE}
knitr::opts_chunk$set(results='hide',
                      eval = FALSE,
                      root.dir = "..",
                      class.source="r_code",
                      class.output="r_output",
                      class.warning="r_warning",
                      class.message="r_warning",
                      class.error="r_error")
```

<br>

-----

## Goals

Process metabarcoding reads, post-adapter removal, including:

- Evaluation of quality
- Quality trimming and filtering
- Error correction and denoising
- Read merging
- Assigning taxonomy

-----

## Before We Get Started

### Notes

- This documented was adapted from [Callahan et al. 2006](https://f1000research.com/articles/5-1492/v2)
  by Matthew Willman, with further edits by Soledad Benitez Ponce and Jelmer Poelstra."
  
- To convert an `Rmd` (R Markdown) file to an R script,
  type the following in an R console:
  `knitr::purl(input="<filename>.Rmd")`.
  (You can download this `Rmd` file by clicking the `Code` button
  in the top-right of this page -- but we will open it at OSC.)

### Start an RStudio Server job at OSC

- Login to OSC at <https://ondemand.osc.edu>.

- Click on `Interactive Apps` (top bar) > `RStudio Server (Owens and Pitzer)`

<p align="center">
<img src=slides/figs_OSC/apps.png width="500">
</p>

- Fill out the form as shown [here](slides/03-OSC.html#rstudio_server_job)

- Once your job has started, click `Launch RStudio Server`.

- In the R console, enter:

  ```sh
  setwd('/fs/project/PAS0471/workshops/2020-12_micro')
  ```

- Then in the Files pane, in the bottom right, click the Gear icon and
  then `Go To Working Directory`, as shown below:
  
<p align="center">
<img src=img/RStudio_go-to-wd.png width="500">
</p>

- Now, click on the `markdown` directory in the Files pane,
  and open the file `06-reads-to-ASV.Rmd`. That's this file!


-----

## Step 1: Getting Started

### Install and load packages

To save time, we have already installed all the necessary R packages at OSC
into a custom library.
To add this library for the current R session:

```{r libload}
.libPaths(new = '/fs/project/PAS0471/.R/4.0/')
```


```{r}
packages <- c("ggplot2", "gridExtra",
              "dada2", "phyloseq", "DECIPHER", "phangorn")
pacman::p_load(char = packages, install = TRUE)
```

### Set the file paths

```{r}
# Dir with input fastq files:
indir <- "processed_data/fastq_trimmed/second_trim/"

# Dir for output:
outdir_base <- 'analysis/ASV_inference/output/'
me <- system("echo $USER", intern = TRUE)
outdir <- file.path(outdir_base, me)
dir.create(outdir, recursive = TRUE)
outdir

# Fasta file with training data:
# (Check for an up-to-date version at <https://benjjneb.github.io/dada2/training.html>)
tax_key <- "analysis/ASV_inference/input/silva_nr99_v138_train_set.fa" 

# File with sample metadata:
sample_names_file <- "metadata/sample_meta.txt"
```

### Assign files to forward and reverse reads 

For forward and reverse fastq filenames which have the following name format: `<SAMPLENAME>_R1_001.fastq` and `<SAMPLENAME>_R2_001.fastq`.

```{r}
fnFs <- sort(list.files(indir, pattern="_R1_001.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(indir, pattern="_R2_001.fastq.gz", full.names = TRUE))
```

### Extract sample names

Extract the sample names from the `.fastq` file names, assuming file names have format: `<SAMPLENAME>_XXX.fastq`.

The resulting `sample.names` should match the SampleID column in the `sample_data_example` file.

```{r}
sample.names <- sapply(strsplit(basename(fnFs), "-V4-V5_"), `[`, 1)
```

-----

## Step 2: QC

### Plot sequence quality data

You can generate and evaluate plots individually:

```{r}
plotQualityProfile(fnFs[1:2])
```

<br>

Or generate pdf files to inspect multiple plots at once:

```{r}
ii <- sample(length(fnFs), 3)
pdf("Error_profiles.pdf")
par(mfrow=c(2,3))
for (i in ii) { print (plotQualityProfile(fnFs[i]) + ggtitle("Fwd")) }
ii <- sample(length(fnRs), 3)
for (i in ii) { print (plotQualityProfile(fnRs[i]) + ggtitle("Rev")) }
dev.off()
```




-----

## Step 3: Trimming

The `truncLen` parameters should be based on sequence quality visualized above.
It is suggested to trim first 10 nucleotides of each read,
empirical observations across Illumina datasets show these positions are likely to contain pathological errors. Trimming length could be different for R1 and R2,
but ideally not too drastically so, in order to maximize merging pairs in subsequent steps.

First, define paths for trimming:

```{r}
filtFs <- file.path(output, "filtered", paste0(sample.names, "_F_filt.fastq"))
filtRs <- file.path(output, "filtered", paste0(sample.names, "_R_filt.fastq"))
```

<br>

Then, perform the trimming:

```{r}
print("filter and Trimming")
Sys.time()
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, trimLeft=10, truncLen=c(250,210), maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=F, compress=FALSE, multithread=TRUE, verbose=TRUE) 
print("done")
Sys.time()
```


-----

## Step 4: Dereplication & Error Training

```{r}
derepFs <- derepFastq(filtFs, verbose = FALSE)
derepRs <- derepFastq(filtRs, verbose = FALSE)

print("Learning errors:")
Sys.time()
errF <- learnErrors(derepFs, multithread = TRUE, verbose = TRUE)
errR <- learnErrors(derepRs, multithread = TRUE, verbose = TRUE)
print("Done.")
Sys.time()

names(derepFs) <- sample.names
names(derepRs) <- sample.names
```

<br>

Plot errors to verify that error rates have been reasonable well-estimated.
Observe the fit between observed error rates (points) and fitted error rates (lines):

```{r}
plotErrors(errF, nominalQ = TRUE)
plotErrors(errR, nominalQ = TRUE)
```

-----

## Step 5: Infer ASVs

We will run the dada algorithm to infer ASVs.

This step is more computationally intensity. Indepedant inference by sample (pool=F) will keep memory requirements linear to sample number. Pooling will increase computation time but improve detection of rare variants seen once or twice in an individual sample but many times across all samples.

```{r}
print("inferring ASVs (running dada algorithm)")
Sys.time()
dadaFs <- dada(derepFs, err = errF, multithread = T, pool = FALSE)
dadaRs <- dada(derepRs, err = errR, multithread = T, pool = FALSE)
print("done")
Sys.time()

#inspect object
dadaFs[[1]]
```

-----

## Step 6: Merge Read Pairs

In this step, we will merge read pairs and generate sequence table of paired ASVs.

```{r}
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs,
                      verbose=TRUE)

seqtab.all <- makeSequenceTable(mergers)
dim(seqtab.all)
```

-----

## Step 7: Summarize Sequences and Remove Chimeras

```{r}
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab.all)))

#If you need to remove sequences of a particular length (e.g. too long):
#seqtab2 <- seqtab[,nchar(colnames(seqtab.all)) %in% seq(250,256)]

# Remove chimeras:
seqtab <- removeBimeraDenovo(seqtab.all,
                             method = "consensus",
                             multithread = TRUE,
                             verbose = TRUE)
dim(seqtab)

# Proportion of removed sequences
sum(seqtab) / sum(seqtab.all)

# Save seqtab file as an RDS object
saveRDS(seqtab, file = file.path(outdir, "seqtab_V4.rds"))
```


-----


## Step 8: Generate a Summary Table

In this step, we will generate a summary table of the number of sequences
processed and outputs of different steps of the pipeline.

This information is generally used to further evaluate characteristics and
quality of the run, sample-to-sample variation,
and resulting sequencing depth for each sample. 

```{r}
getN <- function(x) sum(getUniques(x))

track <- cbind(out,
               sapply(dadaFs, getN),
               sapply(dadaRs, getN),
               sapply(mergers, getN),
               rowSums(seqtab))

# If processing a single sample, remove the sapply calls:
# e.g. replace sapply(dadaFs, getN) with getN(dadaFs)

colnames(track) <- c("input", "filtered", "denoisedF",
                     "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)

outfile_track <- file.path(outdir, 'track.txt')
write.table(track, file = outfile_track)
```

-----

## Step 9: Assign Taxonomy to ASVs

Depending on the marker gene and the data, you will choose the appropriate file for this step. Several files have been formatted for taxony assignments in DADA2 pipeline and are available in the following website: https://benjjneb.github.io/dada2/index.html 

```{r}
print("Assigning taxa to ASVs")
Sys.time()
taxa <- assignTaxonomy(seqtab, tax_key, multithread = TRUE)
colnames(taxa) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")
print("done")
Sys.time()
```

-----

## Step 10: Phylogenetic Tree Estimation

A phylogenetic tree can be estimated for the sequence data you generated. Depending on the number of ASVs recovered and the phylogenetic tree algorithm of choice, this step could take several days. Simpler trees will be less computationally intensive. Depending on the marker gene you are working on, you might choose or not to use this option. 

```{r, eval=FALSE}
#seqtab<- readRDS("seqtab_V4.rds")
seqs <- getSequences(seqtab)
names(seqs) <- seqs # This propagates to the tip labels of the tree. At this stage ASV labels are full ASV sequence
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA, iterations = 5, refinements = 5)

print("Computing pairwise distances from ASVs")
Sys.time()
phang.align <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phang.align)
treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phang.align)
print("done")
Sys.time()

fitGTR <- update(fit, k=4, inv=0.2)
print("done")
Sys.time()

print("Computing likelihood of tree")
Sys.time()
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                      rearrangement = "stochastic", control = pml.control(trace = 0))
print("done")
Sys.time()
```



-----

## Step 11: Generate Output Files

In this step, we will generate output files formatted for downstream analysis in phyloseq

Data will be extracted from the DADA2 outputs and formatted for downstream analysis and data inspection. The last step of this pipeline saves a phyloseq object as an .RDS file and can be imported directly in phyloseq. 

```{r}
# Rename ASV names: 
asv_seqs<- colnames(seqtab)
asv_headers<- vector(dim(seqtab)[2], mode = "character")

for (i in 1:dim(seqtab)[2]) {
	asv_headers[i] <- paste(">ASV", i, sep = "_")
}


# Write ASV fasta for final ASV seqs.
#This fasta file can also be used for phylogenetic tree inference in different platforms
asv_fasta<- c(rbind(asv_headers, asv_seqs))

outfile_fasta <- file.path(outdir, 'ASVs.fa')
write(asv_fasta, outfile_fasta)


# Count table:
asv_tab <- t(seqtab)
row.names(asv_tab) <- sub(">", "", asv_headers)

outfile_tab <- file.path(outdir, 'ASVs_counts.tsv')
write.table(asv_tab, outfile_tab, 
            sep = "\t", quote = FALSE, col.names = NA)


# Tax table:
asv_tax <- taxa
row.names(asv_tax) <- sub(">", "", asv_headers)

outfile_tax <- file.path(outdir, 'ASVs_taxonomy.tsv')
write.table(asv_tax, outfile_tax,
            sep = "\t", quote = FALSE, col.names = NA)


# Read files back in:
count_tab <- read.table(outfile_tab,
                        header = TRUE, row.names = 1,
                        check.names = FALSE, sep = "\t")

tax_tab <- as.matrix(
  read.table(outfile_tax,
             header = TRUE, row.names = 1,
             check.names = FALSE, sep = "\t")
  )


# Load and prepare sample metadata:
sample_df <- read.table(file = sample_names_file,
                        sep = "\t", header = TRUE)
colnames(samdf)[1]<- "SampleID"
rownames(samdf)<- samdf$SampleID


# Build the final phyloseq object:
# We will not do this now, but a phylogenetic tree can also be added to the phyloseq object.
ps <- phyloseq(otu_table(count_tab, taxa_are_rows = TRUE),
               sample_data(sample_df),
               tax_table(tax_tab))

outfile_ps <- file.path(outdir, 'ps_V4.rds')
saveRDS(ps, file = outfile_ps)

print("done")
```

-----

## Resources

- [Callahan et al. 2006: "Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses"](https://f1000research.com/articles/5-1492/v2)
- [`dada2` documentation and tutorials](https://benjjneb.github.io/dada2/index.html )
- [Taxonomic references for `dada2`](https://benjjneb.github.io/dada2/training.html)
- [`cutadapt` documentation and tutorials](https://cutadapt.readthedocs.io/en/stable/index.html)


-----

## Bonus: Submit script as an OSC job

- Script `dada2.sh`:

```sh
#PBS -l walltime=100:00:00
#PBS -l nodes=1:ppn=28
#PBS -A PAS1548

module load gnu/9.1.0
module load mkl/2019.0.3
module load R/3.6.1
cd /fs/scratch/PAS1548/BaseSpace_out/11-21-19/V4_V5/scripts
Rscript dada2_V4.R
```

</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("06-reads-to-ASV.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
