<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title> Workflow part III: Analyzing ASV Data</title>

<script src="site_libs/header-attrs-2.5/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
#rmd-source-code {
  display: none;
}
</style>

<link rel="stylesheet" href="my.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Microbiomics Workshop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Overview
  </a>
</li>
<li>
  <a href="00-setup.html">
    <span class="fa fa-gear"></span>
     
    Setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Lectures:</li>
    <li>
      <a href="01-microbiomics.html">Intro to Amplicon Sequencing</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Lecture/Code-along:</li>
    <li>
      <a href="02-shell.html">Intro to the Shell</a>
    </li>
    <li>
      <a href="03-OSC.html">Intro to OSC</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Workflow Sessions:</li>
    <li>
      <a href="04-data-overview.html">Overview</a>
    </li>
    <li>
      <a href="05-preprocess-reads.html">I: Adapter and Primer Removal</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Lecture/Code-along:</li>
    <li>
      <a href="06-R.html">Intro to R</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Workflow Sessions:</li>
    <li>
      <a href="07-ASV-inference.html">II: ASV Inference</a>
    </li>
    <li>
      <a href="08-postASV-analysis.html">III: ASV Analysis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/mcic-osu/2020-12-microbiomics-workshop.git">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore"><br>Workflow part III:<br>Analyzing ASV Data</h1>

</div>


<p><br></p>
<hr />
<div id="goals" class="section level2">
<h2>Goals</h2>
<p>Analyze microbiome experimental data as a phyloseq object - explore ecological metrics and identify differentially abundant taxa.</p>
<hr />
</div>
<div id="before-we-get-started" class="section level2">
<h2>Before We Get Started</h2>
<div id="notes" class="section level3">
<h3>Notes</h3>
<ul>
<li><p>This is a slightly modified version of a document created by <a href="mailto:willman.18@osu.edu">Matthew Willman</a>.</p></li>
<li><p>To convert an <code>Rmd</code> (R Markdown) file to an R script, type the following in an R console: <code>knitr::purl(input="&lt;filename&gt;.Rmd")</code>. (You can download this <code>Rmd</code> file by clicking the <code>Code</code> button in the top-right of this page – but we will open it at OSC.)</p></li>
</ul>
</div>
<div id="open-this-file-in-rstudio-server-at-osc" class="section level3">
<h3>Open this file in RStudio Server at OSC</h3>
<ul>
<li><p>This assumes you still have an active RStudio Server job at OSC; if not see the <a href="07-ASV-inference.html#start-an-rstudio-server-job-at-osc">instructions on the previous page</a> to start one.</p></li>
<li><p>In the Files pane, in the <code>markdown</code> directory, open the file <code>08-postASV-analysis.Rmd</code>. That’s this file!</p></li>
</ul>
<hr />
</div>
</div>
<div id="getting-started" class="section level2">
<h2>Getting Started</h2>
<div id="install-and-load-packages" class="section level3">
<h3>Install and Load Packages</h3>
<p>Add our custom library again if you restarted your R session:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">.libPaths</span>(<span class="dt">new =</span> <span class="st">&#39;/fs/project/PAS0471/.R/4.0/&#39;</span>)</span></code></pre></div>
<p>And load the necessary packages:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>packages &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;tidyverse&#39;</span>, <span class="st">&#39;vegan&#39;</span>, <span class="st">&#39;phyloseq&#39;</span>, <span class="st">&#39;decontam&#39;</span>, <span class="st">&#39;ape&#39;</span>,</span>
<span id="cb2-2"><a href="#cb2-2"></a>              <span class="st">&#39;DESeq2&#39;</span>, <span class="st">&#39;microbiome&#39;</span>, <span class="st">&#39;metagenomeSeq&#39;</span>, <span class="st">&#39;remotes&#39;</span>,</span>
<span id="cb2-3"><a href="#cb2-3"></a>              <span class="st">&#39;ampvis2&#39;</span>, <span class="st">&#39;breakaway&#39;</span>)</span>
<span id="cb2-4"><a href="#cb2-4"></a>pacman<span class="op">::</span><span class="kw">p_load</span>(<span class="dt">char =</span> packages)</span>
<span id="cb2-5"><a href="#cb2-5"></a></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co"># If you wanted to install and load these packages yourself,</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co"># first make sure you have the pacman package installed:</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co">## install.packages(&#39;pacman&#39;)</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co"># Then, the code above would work except for the last two packages,</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co"># which are available from Github only. To install these, you would run:</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">## remotes::install_github(&quot;MadsAlbertsen/ampvis2&quot;)</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co">## remotes::install_github(&quot;adw96/breakaway&quot;)</span></span></code></pre></div>
</div>
<div id="set-directories" class="section level3">
<h3>Set Directories</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># Dir with input files:</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>indir &lt;-<span class="st"> &#39;analysis/ASV_inference/&#39;</span></span>
<span id="cb3-3"><a href="#cb3-3"></a></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co"># Dir for output:</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>outdir &lt;-<span class="st"> &#39;analysis/postASV_analysis/&#39;</span></span></code></pre></div>
</div>
<div id="load-the-data" class="section level3">
<h3>Load the Data</h3>
<p>This session starts with a <em>phyloseq</em> object similar to the one generated using <em>DADA2</em> in the <a href="06-reads-to-ASV.Rmd">previous session</a>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>ps_raw &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">file.path</span>(indir, <span class="st">&#39;ps_16S_V4_withtree.rds&#39;</span>))</span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co"># This would load the object from the previous session, which doesn&#39;t contain a tree:</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co"># ps_raw &lt;- readRDS(file.path(indir, &#39;ps_V4.rds&#39;))</span></span></code></pre></div>
<p>Our <em>phyloseq</em> object is made up of the following components (slots):</p>
<ul>
<li><code>otu_table</code></li>
<li><code>sample_data</code></li>
<li><code>tax_table</code></li>
<li><code>phy_tree</code></li>
</ul>
<p>Let’s have a look at the object:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>ps_raw</span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co"># And the assigned taxon names:</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="kw">taxa_names</span>(ps_raw)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<p>As we can see, the taxon names are currently the associated sequences. We can create a new <em>phyloseq</em> component to store these sequences, then rename the ASVs so something shorter (“ASV_1”, “ASV_2”, etc.).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>dna &lt;-<span class="st"> </span>Biostrings<span class="op">::</span><span class="kw">DNAStringSet</span>(<span class="kw">taxa_names</span>(ps_raw))</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw">names</span>(dna) &lt;-<span class="st"> </span><span class="kw">taxa_names</span>(ps_raw)</span>
<span id="cb6-3"><a href="#cb6-3"></a></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co"># Merge sequence object into the phyloseq object:</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>ps_raw &lt;-<span class="st"> </span><span class="kw">merge_phyloseq</span>(ps_raw, dna)</span>
<span id="cb6-6"><a href="#cb6-6"></a>ps_raw</span>
<span id="cb6-7"><a href="#cb6-7"></a></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co"># Rename ASVs:</span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="kw">taxa_names</span>(ps_raw) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;ASV&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">ntaxa</span>(ps_raw), <span class="dt">sep =</span> <span class="st">&quot;_&quot;</span>)</span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="kw">taxa_names</span>(ps_raw)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<hr />
</div>
</div>
<div id="filter-taxa" class="section level2">
<h2>Filter taxa</h2>
<div id="identify-and-remove-contaminants" class="section level3">
<h3>Identify and Remove Contaminants</h3>
<p>It is possible to introduce contaminating microbes during sample preparation. Before analyzing the data, we will identify and remove probable contaminants using the <a href="https://benjjneb.github.io/decontam/vignettes/decontam_intro.html">decontam package</a>.</p>
<p>In this case, we will define a conataminant as an ASV whose abundance correlates with DNA concentration (post-PCR). Our assumption here is that each soil taxon’s abundance should be independant of DNA concentration. However, if we were to spike each sample with a contaminant, that contaminant would show a greater abundance in samples with lower DNA concentrations. To do this, we need DNA concentration measured after PCR and before pooling of the samples.</p>
<p>These data are in column 7 of the sample data component:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">head</span>(<span class="kw">sample_data</span>(ps_raw))</span></code></pre></div>
<p>MCIC measured our DNA concentration by comparing band intensities to a single reference band. MCIC’s DNA concentration measurements range from 0 to 1 (the reference). Zero in this case is coded as <code>NA</code>. The <em>decontam</em> package can’t handle NA’s or 0’s, so we need to change them to a small value (e.g. 0.01).</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># First, we identify the NAs in the DNA measurement column:</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>NAs &lt;-<span class="st"> </span><span class="kw">is.na</span>(<span class="kw">sample_data</span>(ps_raw)<span class="op">$</span>PCR_prod_V4.V5)</span>
<span id="cb8-3"><a href="#cb8-3"></a></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co"># Then, we replace NAs with 0.01:</span></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="kw">sample_data</span>(ps_raw)<span class="op">$</span>PCR_prod_V4.V5[NAs] &lt;-<span class="st"> </span><span class="fl">0.01</span></span></code></pre></div>
<p>We will use <code>decontam::isContaminant(method = "frequency")</code> to test each taxon against the null hypothesis: abundance is not associated with DNA concentration. Because the DNA concentration measurements are not very accurate, and we want to make sure to remove possible contaminants, we use a relaxed p-value (0.2) to reject the null hypothesis:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>contam_df_freq &lt;-<span class="st"> </span><span class="kw">isContaminant</span>(ps_raw,</span>
<span id="cb9-2"><a href="#cb9-2"></a>                                <span class="dt">method =</span> <span class="st">&quot;frequency&quot;</span>,</span>
<span id="cb9-3"><a href="#cb9-3"></a>                                <span class="dt">conc =</span> <span class="st">&quot;PCR_prod_V4.V5&quot;</span>,</span>
<span id="cb9-4"><a href="#cb9-4"></a>                                <span class="dt">threshold =</span> <span class="fl">0.2</span>)</span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="kw">head</span>(contam_df_freq)</span></code></pre></div>
<p>As we can see, the top 3 ASVs are identified as probable contaminants. How many contaminants were identified (<code>TRUE</code> in the <code>contaminant</code> column) in total?</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">table</span>(contam_df_freq<span class="op">$</span>contaminant)</span></code></pre></div>
<p>What are the most abundant contaminant taxa?</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="co"># Get abundance ranks for contaminant ASVs:</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="kw">head</span>(<span class="kw">which</span>(contam_df_freq<span class="op">$</span>contaminant))</span>
<span id="cb11-3"><a href="#cb11-3"></a></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="co"># Check which taxa are contaminants:</span></span>
<span id="cb11-5"><a href="#cb11-5"></a>ps_contam &lt;-<span class="st"> </span><span class="kw">prune_taxa</span>(contam_df_freq<span class="op">$</span>contaminant, ps_raw)</span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="kw">head</span>(<span class="kw">tax_table</span>(ps_contam))</span></code></pre></div>
<p>Let’s take a look at our tested associations for several taxa:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw">plot_frequency</span>(ps_raw,</span>
<span id="cb12-2"><a href="#cb12-2"></a>               <span class="kw">taxa_names</span>(ps_raw)[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">11</span>)],</span>
<span id="cb12-3"><a href="#cb12-3"></a>               <span class="dt">conc =</span> <span class="st">&quot;PCR_prod_V4.V5&quot;</span>) <span class="op">+</span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;DNA Concentration (relative to single marker)&quot;</span>)</span></code></pre></div>
<p>As we can see, abundance of ASVs 1, 2, 3, and 11 are particularly high in samples with low DNA concentration.</p>
<p>Which samples have highest abundances of the contaminants?</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>asv1 &lt;-<span class="st"> </span><span class="kw">otu_table</span>(ps_raw)[, <span class="dv">1</span>]</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="kw">head</span>(asv1[<span class="kw">order</span>(asv1, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)])</span>
<span id="cb13-3"><a href="#cb13-3"></a></span>
<span id="cb13-4"><a href="#cb13-4"></a>asv2 &lt;-<span class="st"> </span><span class="kw">otu_table</span>(ps_raw)[, <span class="dv">2</span>]</span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="kw">head</span>(asv2[<span class="kw">order</span>(asv2, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)])</span></code></pre></div>
<p>How do these contaminants overlap with ASVs present in the negative control?</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw">otu_table</span>(ps_raw)[<span class="dv">70</span><span class="op">:</span><span class="dv">75</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">11</span>]</span></code></pre></div>
<p>ASVs have high read counts in the extraction negative control. Thus, we can be fairly certain these are contaminants and not representative of our experimental samples.</p>
<p>Remove the contaminants that we identified:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>ps_noncontam &lt;-<span class="st"> </span><span class="kw">prune_taxa</span>(<span class="op">!</span>contam_df_freq<span class="op">$</span>contaminant, ps_raw)</span>
<span id="cb15-2"><a href="#cb15-2"></a></span>
<span id="cb15-3"><a href="#cb15-3"></a>ps_noncontam</span></code></pre></div>
<p>What proportion of our count data were removed as contaminants?</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>pre &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">sample_sums</span>(ps_raw))</span>
<span id="cb16-2"><a href="#cb16-2"></a>post &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">sample_sums</span>(ps_noncontam))</span>
<span id="cb16-3"><a href="#cb16-3"></a></span>
<span id="cb16-4"><a href="#cb16-4"></a>(pre<span class="op">-</span>post) <span class="op">/</span><span class="st"> </span>pre</span></code></pre></div>
</div>
<div id="remove-non-bacterial-non-archaeal-asvs" class="section level3">
<h3>Remove non-bacterial, non-archaeal ASVs</h3>
<p>The V4 region of 16S rRNA is conserved within certain bacteria, archaea, chloroplasts, mitochondria, and eukaryotes. Since we are only interested in bacteria and archaea here, we’ll need to remove other taxa from our object.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># Create ps subsets for chloroplast, mitochondria, and eukaryotes:</span></span>
<span id="cb17-2"><a href="#cb17-2"></a>chlr &lt;-<span class="st"> </span><span class="kw">subset_taxa</span>(ps_noncontam, Order <span class="op">==</span><span class="st"> &quot;Chloroplast&quot;</span>)</span>
<span id="cb17-3"><a href="#cb17-3"></a>mit &lt;-<span class="st"> </span><span class="kw">subset_taxa</span>(ps_noncontam, Family <span class="op">==</span><span class="st"> &quot;Mitochondria&quot;</span>)</span>
<span id="cb17-4"><a href="#cb17-4"></a>euk &lt;-<span class="st"> </span><span class="kw">subset_taxa</span>(ps_noncontam, Kingdom <span class="op">==</span><span class="st"> &quot;Eukaryota&quot;</span>)</span>
<span id="cb17-5"><a href="#cb17-5"></a></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="co"># Get all taxon IDs that we want to remove:</span></span>
<span id="cb17-7"><a href="#cb17-7"></a>bad_taxa &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">taxa_names</span>(chlr), <span class="kw">taxa_names</span>(mit), <span class="kw">taxa_names</span>(euk))</span>
<span id="cb17-8"><a href="#cb17-8"></a></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="co"># Get all taxon IDs that we want to keep:</span></span>
<span id="cb17-10"><a href="#cb17-10"></a>all_taxa &lt;-<span class="st"> </span><span class="kw">taxa_names</span>(ps_noncontam)</span>
<span id="cb17-11"><a href="#cb17-11"></a>good_taxa &lt;-<span class="st"> </span>all_taxa[<span class="op">!</span>(all_taxa <span class="op">%in%</span><span class="st"> </span>bad_taxa)]</span>
<span id="cb17-12"><a href="#cb17-12"></a></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="co"># Subset phyloseq object:</span></span>
<span id="cb17-14"><a href="#cb17-14"></a>ps_bac_arc &lt;-<span class="st"> </span><span class="kw">prune_taxa</span>(good_taxa, ps_noncontam)</span></code></pre></div>
<p>What proportion of ASVs were kept?</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>pre &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">sample_sums</span>(ps_noncontam))</span>
<span id="cb18-2"><a href="#cb18-2"></a>post &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">sample_sums</span>(ps_bac_arc))</span>
<span id="cb18-3"><a href="#cb18-3"></a>post <span class="op">/</span><span class="st"> </span>pre</span></code></pre></div>
<p>Finally, we’ll also remove non-experimental samples (e.g. negative controls). Note, though, that non-experimental samples are important for evaluating the quality of your reagents, consistency across sequencing runs, and contamination.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>ps &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(ps_bac_arc, Experiment <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>)</span>
<span id="cb19-2"><a href="#cb19-2"></a></span>
<span id="cb19-3"><a href="#cb19-3"></a>ps</span></code></pre></div>
<hr />
</div>
</div>
<div id="filter-samples" class="section level2">
<h2>Filter samples</h2>
<p>After having filtered out unwanted taxa, we now need to filter out uninformative samples – those with low total taxon counts.</p>
<p>First, how many counts do we have for each sample?</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>sums &lt;-<span class="st"> </span><span class="kw">sample_sums</span>(ps)</span>
<span id="cb20-2"><a href="#cb20-2"></a>sums[<span class="kw">order</span>(sums)]</span>
<span id="cb20-3"><a href="#cb20-3"></a></span>
<span id="cb20-4"><a href="#cb20-4"></a><span class="co"># Lowest total count:</span></span>
<span id="cb20-5"><a href="#cb20-5"></a><span class="kw">min</span>(sums)</span>
<span id="cb20-6"><a href="#cb20-6"></a></span>
<span id="cb20-7"><a href="#cb20-7"></a><span class="co"># Sample with lowest counts:</span></span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="kw">names</span>(sums[<span class="kw">order</span>(sums)][<span class="dv">1</span>])</span></code></pre></div>
<p>To remove uninformative samples, we will only keep those with over 1,000 counts. We will also remove a technical replicate, though note that technical replicates are important for quality control of runs and for understanding the reproducibility of the methodology.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co"># Remove samples with low counts:</span></span>
<span id="cb21-2"><a href="#cb21-2"></a>ps &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(ps, <span class="kw">sample_sums</span>(ps) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1000</span>)</span>
<span id="cb21-3"><a href="#cb21-3"></a></span>
<span id="cb21-4"><a href="#cb21-4"></a><span class="co"># Remove the technical replicate:</span></span>
<span id="cb21-5"><a href="#cb21-5"></a>ps &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(ps, SampleID <span class="op">!=</span><span class="st"> &quot;501S4&quot;</span>)</span></code></pre></div>
<p>Let’s take a look at our current <code>phyloseq</code> object:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="kw">sample_data</span>(ps)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="kw">head</span>(<span class="kw">t</span>(<span class="kw">otu_table</span>(ps)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]))</span>
<span id="cb22-4"><a href="#cb22-4"></a></span>
<span id="cb22-5"><a href="#cb22-5"></a><span class="kw">head</span>(<span class="kw">tax_table</span>(ps))</span></code></pre></div>
<hr />
</div>
<div id="normalization" class="section level2">
<h2>Normalization</h2>
<p>Normalization is currently a much-discussed issue of microbiome studies. Differences in read depth between samples often need to be corrected before analysis. Several normalization methods have been proposed, and no single method is perfect. It may be that the most appropriate method depends on the analysis.</p>
<p>In this tutorial, we will use:</p>
<ul>
<li><p>Proportion-normalized data to estimate ecological metrics.</p></li>
<li><p>Stabilizing transformation normalized data to identify differentially abundant taxa.</p></li>
</ul>
<p>However, for your reference, we will also present code for two other normalization methods: rarefaction and cumulative sum scaling.</p>
<p>For more details, read <a href="https://dx.plos.org/10.1371/journal.pcbi.1003531">McMurdie and Holmes 2014</a> and <a href="http://doi.wiley.com/10.1111/2041-210X.13115">McKnight et al. 2019</a>.</p>
<hr />
<div id="proportion-normalization" class="section level3">
<h3>Proportion Normalization</h3>
<p>Proportion normalization involves dividing each OTU count by the total sum for each sample. The resulting count data will add up to 1 (100%) for each sample.</p>
<p>The <code>microbiome::transform</code> function can be used to easily normalize count data as proportions in a <em>phyloseq</em> object:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="co"># Proportion normalization:</span></span>
<span id="cb23-2"><a href="#cb23-2"></a>ps_prop &lt;-<span class="st"> </span><span class="kw">transform</span>(ps, <span class="st">&quot;compositional&quot;</span>)</span>
<span id="cb23-3"><a href="#cb23-3"></a></span>
<span id="cb23-4"><a href="#cb23-4"></a><span class="co"># Have a look at the resulting OTU table:</span></span>
<span id="cb23-5"><a href="#cb23-5"></a><span class="kw">otu_table</span>(ps_prop)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb23-6"><a href="#cb23-6"></a></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="co"># Check what the sums for each sample are now:</span></span>
<span id="cb23-8"><a href="#cb23-8"></a><span class="kw">head</span>(<span class="kw">sample_sums</span>(ps_prop))</span></code></pre></div>
<hr />
</div>
<div id="variance-stabilizing-transformation" class="section level3">
<h3>Variance stabilizing transformation</h3>
<p>This transformation method, available in <a href="https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html">DESeq2</a>, normalizes data with respect to library size as well as dispersion.<br />
Type <code>?vst</code> or read <a href="http://dx.doi.org/10.1186/gb-2010-11-10-r106">Anders and Huber 2010</a> for more information.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co"># First we convert the phyloseq object to a DESeq object:</span></span>
<span id="cb24-2"><a href="#cb24-2"></a>dds &lt;-<span class="st"> </span><span class="kw">phyloseq_to_deseq2</span>(ps, <span class="op">~</span><span class="dv">1</span>)</span>
<span id="cb24-3"><a href="#cb24-3"></a></span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="co"># Then we estimate the size factors:</span></span>
<span id="cb24-5"><a href="#cb24-5"></a>dds &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(dds)</span>
<span id="cb24-6"><a href="#cb24-6"></a></span>
<span id="cb24-7"><a href="#cb24-7"></a><span class="co"># Now can do the transformation:</span></span>
<span id="cb24-8"><a href="#cb24-8"></a>dds_vst &lt;-<span class="st"> </span><span class="kw">vst</span>(dds, <span class="dt">blind =</span> <span class="ot">TRUE</span>)</span>
<span id="cb24-9"><a href="#cb24-9"></a></span>
<span id="cb24-10"><a href="#cb24-10"></a><span class="co"># Have a look at the resulting counts:</span></span>
<span id="cb24-11"><a href="#cb24-11"></a>vst_counts &lt;-<span class="st"> </span><span class="kw">assay</span>(dds_vst)</span>
<span id="cb24-12"><a href="#cb24-12"></a><span class="kw">t</span>(vst_counts)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<hr />
</div>
<div id="rarefaction" class="section level3">
<h3>Rarefaction</h3>
<p>Rarefaction can be used to subset data such that the library depth is the same for each sample. Because sampling of the data is random, rarefaction can account for an effect of total read count on taxa richness. However, rarefaction is no longer considered to be a good way to normalize amplicon sequencing data (see <a href="https://dx.plos.org/10.1371/journal.pcbi.1003531">McMurdie &amp; Holmes 2014</a>).</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>ps_rarefied &lt;-<span class="st"> </span><span class="kw">rarefy_even_depth</span>(ps,</span>
<span id="cb25-2"><a href="#cb25-2"></a>                                 <span class="dt">rngseed =</span> <span class="dv">1</span>,</span>
<span id="cb25-3"><a href="#cb25-3"></a>                                 <span class="dt">sample.size =</span> <span class="kw">min</span>(<span class="kw">sample_sums</span>(ps)),</span>
<span id="cb25-4"><a href="#cb25-4"></a>                                 <span class="dt">replace =</span> <span class="ot">FALSE</span>)</span>
<span id="cb25-5"><a href="#cb25-5"></a></span>
<span id="cb25-6"><a href="#cb25-6"></a><span class="co"># Have a look at the result:</span></span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="kw">otu_table</span>(ps_rarefied)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb25-8"><a href="#cb25-8"></a></span>
<span id="cb25-9"><a href="#cb25-9"></a><span class="kw">head</span>(<span class="kw">sample_sums</span>(ps_rarefied))</span></code></pre></div>
<p>Compare these sample sums to the non-rarified data sums:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="kw">head</span>(<span class="kw">sample_sums</span>(ps))</span></code></pre></div>
<hr />
</div>
<div id="cumulative-sum-scaling" class="section level3">
<h3>Cumulative Sum Scaling</h3>
<p>The <em>metagenomeSeq</em> Cumulative Sum Scaling (CSS) normalization is another option developed for microbiome data. For more information, read <a href="http://www.nature.com/articles/nmeth.2658">Paulson et al. 2013</a>.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="co"># Convert the phyloseq object to a metagenomeseq object:</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>mgs_css &lt;-<span class="st"> </span><span class="kw">phyloseq_to_metagenomeSeq</span>(ps)</span>
<span id="cb27-3"><a href="#cb27-3"></a></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="co"># Perform the Cumulative Sum Scaling:</span></span>
<span id="cb27-5"><a href="#cb27-5"></a>mgs_css &lt;-<span class="st"> </span><span class="kw">cumNorm</span>(mgs_css)</span>
<span id="cb27-6"><a href="#cb27-6"></a></span>
<span id="cb27-7"><a href="#cb27-7"></a><span class="co"># Extract the counts and add them to a separate phyloseq object:</span></span>
<span id="cb27-8"><a href="#cb27-8"></a>css_counts &lt;-<span class="st"> </span><span class="kw">MRcounts</span>(mgs_css, <span class="dt">norm =</span> <span class="ot">TRUE</span>)</span>
<span id="cb27-9"><a href="#cb27-9"></a>ps_css &lt;-<span class="st"> </span>ps</span>
<span id="cb27-10"><a href="#cb27-10"></a><span class="kw">otu_table</span>(ps_css) &lt;-<span class="st"> </span><span class="kw">otu_table</span>(<span class="kw">t</span>(css_counts), <span class="dt">taxa_are_rows =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>Now lets compare the original data to the CSS normalized data:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw">otu_table</span>(ps)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb28-2"><a href="#cb28-2"></a><span class="kw">head</span>(<span class="kw">sample_sums</span>(ps))</span>
<span id="cb28-3"><a href="#cb28-3"></a></span>
<span id="cb28-4"><a href="#cb28-4"></a><span class="kw">otu_table</span>(ps_css)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb28-5"><a href="#cb28-5"></a><span class="kw">head</span>(<span class="kw">sample_sums</span>(ps_css))</span></code></pre></div>
<hr />
</div>
</div>
<div id="plot-abundances" class="section level2">
<h2>Plot abundances</h2>
<p>In this and following the examples we will use the <strong>proportions</strong> data.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="co"># Subset to the T1 treatment</span></span>
<span id="cb29-2"><a href="#cb29-2"></a>T1 &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(ps_prop, Treatment <span class="op">==</span><span class="st"> &quot;T1&quot;</span>)</span>
<span id="cb29-3"><a href="#cb29-3"></a></span>
<span id="cb29-4"><a href="#cb29-4"></a><span class="co"># Agglomerate taxa to phylum level:</span></span>
<span id="cb29-5"><a href="#cb29-5"></a>T1_phylum &lt;-<span class="st"> </span><span class="kw">tax_glom</span>(T1, <span class="dt">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="dt">NArm =</span> <span class="ot">FALSE</span>)</span>
<span id="cb29-6"><a href="#cb29-6"></a></span>
<span id="cb29-7"><a href="#cb29-7"></a><span class="co"># Remove uncommon phyla:</span></span>
<span id="cb29-8"><a href="#cb29-8"></a>T1_phylum &lt;-<span class="st"> </span><span class="kw">subset_taxa</span>(T1_phylum, <span class="kw">taxa_sums</span>(T1_phylum) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.1</span>)</span>
<span id="cb29-9"><a href="#cb29-9"></a></span>
<span id="cb29-10"><a href="#cb29-10"></a><span class="co"># Plot:</span></span>
<span id="cb29-11"><a href="#cb29-11"></a><span class="kw">plot_bar</span>(T1_phylum, <span class="dt">fill =</span> <span class="st">&quot;Phylum&quot;</span>) <span class="op">+</span></span>
<span id="cb29-12"><a href="#cb29-12"></a><span class="st">   </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Timepoint, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>, <span class="dt">nrow =</span> <span class="dv">1</span>)</span></code></pre></div>
<hr />
</div>
<div id="alpha-diversity" class="section level2">
<h2>Alpha diversity</h2>
<p>Alpha diversity measures taxonomic diversity within a single population. Measures of alpha diversity include taxonomic richness (i.e. number of taxa), and indices which combine taxonomic richness with some measure of evenness.</p>
<p><em>Note:</em> Many of these methods are greatly influenced by singleton data (i.e. taxa represented by a single count), meaning they may be unreliable if your data excludes singletons. DADA2 does not output singletons if it is run individually on each sample. Thus, if using dada2 to infer ASVs, it is advisable to use <code>pool = TRUE</code> when running the dada algorithm. This will allow calling per-sample singletons, but not per-study singletons.</p>
<div id="taxonomic-richness-rarefaction-plot" class="section level3">
<h3>Taxonomic richness: Rarefaction plot</h3>
<p>Our goal in defining richness is to determine the number of unique taxa within each population (in our case, each plot). This is difficult, as our samples contain only a portion of the total richness within our plots. As sequence depth increases for each sample, so does the number of taxa. This can be illustrated with a rarefaction plot.</p>
<p>To make a rarefaction plot, we draw random samples from our data and count the number of unique ASVs as samples are drawn. The resulting rarefaction curve is expected to rise quickly then plateu as the most abundant taxa are represented. We can make a quick rarefaction curve plot directly from our <code>phyloseq</code> object of all samples using the <a href="https://peat-clark.github.io/BIO381/veganTutorial.html"><code>vegan</code> package</a>:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="kw">rarecurve</span>(<span class="kw">otu_table</span>(ps), <span class="dt">step =</span> <span class="dv">500</span>, <span class="dt">xlab =</span> <span class="st">&quot;Sample Size&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Taxa&quot;</span>)</span></code></pre></div>
<p>We can also split rarefaction curves by group using the <a href="https://madsalbertsen.github.io/ampvis2/articles/ampvis2.html"><code>ampvis2</code> package</a>.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>metadata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">sample_data</span>(ps), <span class="dt">check.names =</span> <span class="ot">FALSE</span>)</span>
<span id="cb31-2"><a href="#cb31-2"></a></span>
<span id="cb31-3"><a href="#cb31-3"></a>asv_table &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">t</span>(<span class="kw">otu_table</span>(ps)),</span>
<span id="cb31-4"><a href="#cb31-4"></a>                       <span class="kw">tax_table</span>(ps),</span>
<span id="cb31-5"><a href="#cb31-5"></a>                       <span class="dt">check.names =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p><code>ampvis2</code> requires a <code>Species</code> column, which must be added to our data. If your table already has a <code>Species</code> column, skip this step:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>asv_table<span class="op">$</span>Species &lt;-<span class="st"> </span><span class="ot">NA</span></span></code></pre></div>
<p>Load an <code>ampvis2</code> object and calculate rarefaction:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>ps3 &lt;-<span class="st"> </span><span class="kw">amp_load</span>(asv_table, metadata)</span>
<span id="cb33-2"><a href="#cb33-2"></a></span>
<span id="cb33-3"><a href="#cb33-3"></a>rar_plot &lt;-<span class="st"> </span><span class="kw">amp_rarecurve</span>(ps3,</span>
<span id="cb33-4"><a href="#cb33-4"></a>                          <span class="dt">stepsize =</span> <span class="dv">100</span>,</span>
<span id="cb33-5"><a href="#cb33-5"></a>                          <span class="dt">facet_by =</span> <span class="st">&quot;Timepoint&quot;</span>,</span>
<span id="cb33-6"><a href="#cb33-6"></a>                          <span class="dt">color_by =</span> <span class="st">&quot;Treatment&quot;</span>) <span class="op">+</span></span>
<span id="cb33-7"><a href="#cb33-7"></a><span class="st">   </span><span class="kw">ylab</span>(<span class="st">&quot;Number of observed ASVs&quot;</span>)</span>
<span id="cb33-8"><a href="#cb33-8"></a></span>
<span id="cb33-9"><a href="#cb33-9"></a>rar_plot</span></code></pre></div>
<p>As we can see, as read depth increases for each sample, so does taxonomic richness. In all samples, increased sampling would result in increased richness. Thus, our observed sample richness is lower than the true population richness for every plot.</p>
<p>Further, sequence depth and total richness appear to be correlated across samples:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>dp_by_asv &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">ct_sums =</span> <span class="kw">sample_sums</span>(ps),</span>
<span id="cb34-2"><a href="#cb34-2"></a>                        <span class="dt">asvs =</span> <span class="kw">rowSums</span>(<span class="kw">otu_table</span>(ps) <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>))</span>
<span id="cb34-3"><a href="#cb34-3"></a></span>
<span id="cb34-4"><a href="#cb34-4"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> dp_by_asv) <span class="op">+</span></span>
<span id="cb34-5"><a href="#cb34-5"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x =</span>  ct_sums, <span class="dt">y =</span> asvs)) <span class="op">+</span></span>
<span id="cb34-6"><a href="#cb34-6"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Sequence depth&quot;</span>, <span class="dt">y=</span> <span class="st">&quot;ASV count&quot;</span>) <span class="op">+</span></span>
<span id="cb34-7"><a href="#cb34-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span></code></pre></div>
<p>We can test whether this correlation is significant:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>cor &lt;-<span class="st"> </span><span class="kw">cor.test</span>(dp_by_asv<span class="op">$</span>ct_sums, dp_by_asv<span class="op">$</span>asvs, <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb35-2"><a href="#cb35-2"></a>cor</span></code></pre></div>
<p>Here we observe a significant correlation between read count and observed taxonomic richness.</p>
<p>Thus, a sample from a population with high actual diversity, but low read depth, could have low observed richness. Fortunately, there are methods to estimate the number of unobserved taxa. For an overview of these methods, read <a href="http://www.annualreviews.org/doi/10.1146/annurev-statistics-022513-115654">Bunge et al. 2014</a>.</p>
</div>
<div id="frequency-counts" class="section level3">
<h3>Frequency counts</h3>
<p>To illustrate the difficulty of estimating population richness and importance of singletons to alpha diversity metrics, we’ll look at the frequency count distribution for our first sample: 101-S1.</p>
<p>We’ll count the number of singletons, doubletons, etc., then plot frequency as a function of count.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="co"># Get the frequency count distribution for the 1st sample:</span></span>
<span id="cb36-2"><a href="#cb36-2"></a>frequencytablelist &lt;-<span class="st"> </span><span class="kw">build_frequency_count_tables</span>(<span class="kw">otu_table</span>(ps))</span>
<span id="cb36-3"><a href="#cb36-3"></a>freq_ct &lt;-<span class="st"> </span>frequencytablelist[[<span class="dv">1</span>]]</span>
<span id="cb36-4"><a href="#cb36-4"></a><span class="kw">colnames</span>(freq_ct) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Frequency&quot;</span>, <span class="st">&quot;Count&quot;</span>)</span>
<span id="cb36-5"><a href="#cb36-5"></a></span>
<span id="cb36-6"><a href="#cb36-6"></a><span class="co"># Plot:</span></span>
<span id="cb36-7"><a href="#cb36-7"></a><span class="kw">ggplot</span>(freq_ct, <span class="kw">aes</span>(Frequency, Count)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb36-8"><a href="#cb36-8"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb36-9"><a href="#cb36-9"></a><span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">600</span>, <span class="dt">by =</span> <span class="dv">50</span>)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb36-10"><a href="#cb36-10"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;101-S1&quot;</span>) <span class="op">+</span></span>
<span id="cb36-11"><a href="#cb36-11"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span></code></pre></div>
<p>We can see there are ~3,200 taxa represented as singletons, and ~1,600 taxa as doubletons. This is a high proportion of the total number taxa are observed in this sample:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="co"># Total number of taxa:</span></span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="kw">sum</span>(freq_ct<span class="op">$</span>Count)</span>
<span id="cb37-3"><a href="#cb37-3"></a></span>
<span id="cb37-4"><a href="#cb37-4"></a><span class="co"># Proportion of singletons</span></span>
<span id="cb37-5"><a href="#cb37-5"></a>freq_ct<span class="op">$</span>Count[<span class="dv">1</span>] <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(freq_ct<span class="op">$</span>Count)</span></code></pre></div>
<p>How many more taxa might we observe if we increased our sampling?</p>
</div>
<div id="diversity-indices" class="section level3">
<h3>Diversity Indices</h3>
<p>We can use several indices to explore how evenly species are distributed within each sample. Note that ‘Observed’ is the sample taxonomic richness. The other three indices combine richness and abundance data for each taxon.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="kw">plot_richness</span>(ps,</span>
<span id="cb38-2"><a href="#cb38-2"></a>              <span class="dt">x =</span> <span class="st">&quot;Timepoint&quot;</span>,</span>
<span id="cb38-3"><a href="#cb38-3"></a>              <span class="dt">shape =</span> <span class="st">&quot;Treatment&quot;</span>,</span>
<span id="cb38-4"><a href="#cb38-4"></a>              <span class="dt">color =</span> <span class="st">&quot;Treatment&quot;</span>, </span>
<span id="cb38-5"><a href="#cb38-5"></a>              <span class="dt">measures =</span> <span class="kw">c</span>(<span class="st">&quot;Observed&quot;</span>, <span class="st">&quot;Shannon&quot;</span>, <span class="st">&quot;Simpson&quot;</span>, <span class="st">&quot;InvSimpson&quot;</span>))</span></code></pre></div>
<p>Notice that several T2:S3 samples have low measures, suggesting they have low diversity (i.e. abundances are dominated by a few taxa) compared to the other samples.</p>
<hr />
</div>
</div>
<div id="beta-diversity" class="section level2">
<h2>Beta diversity</h2>
<p>Beta diversity measures differences in microbial compositions between populations. If available (use ps_16S_V4_withtree.rds), we can use a phylogenetic tree and otu table to estimate the <a href="https://doi.org/10.1128/AEM.71.12.8228-8235.2005">UniFrac</a> for each sample pair.</p>
<p>The result is a distance matrix with height and width equal to the number of samples. UniFrac may be <a href="https://doi.org/10.1128/AEM.01996-06">weighted or unweighted</a>. Selection of weighted or unweighted will depend on the question you want to answer. Unweighted UniFrac uses only presence or absence of a taxon from the otu table (i.e. 1000 will be treated the same as 1). It is appropriate if you want to test qualitative taxa differences between samples. Weighted UniFrac uses presence as well as quantity data (i.e. 1 &lt; 1000).</p>
<p>We are interested to know what changes to taxon abundance may be driven by our treatments, so we will use weighted UniFrac.</p>
<div id="ordination" class="section level3">
<h3>Ordination</h3>
<p>First we’ll make a weighted UniFrac distance matrix. For UniFrac, you require a tree as part of your <em>phyloseq</em> object.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="kw">set.seed</span>(<span class="dv">100</span>)</span>
<span id="cb39-2"><a href="#cb39-2"></a>wUF &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">distance</span>(ps_prop, <span class="dt">method =</span> <span class="st">&quot;wunifrac&quot;</span>)</span>
<span id="cb39-3"><a href="#cb39-3"></a></span>
<span id="cb39-4"><a href="#cb39-4"></a><span class="kw">as.matrix</span>(wUF)[<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</span>
<span id="cb39-5"><a href="#cb39-5"></a></span>
<span id="cb39-6"><a href="#cb39-6"></a><span class="co"># Note: If your *phyloseq* object lacks a phylogenetic tree,</span></span>
<span id="cb39-7"><a href="#cb39-7"></a><span class="co"># you could use a dissimilarity index that does not require one, such as Bray-Curtis:</span></span>
<span id="cb39-8"><a href="#cb39-8"></a><span class="co"># bray &lt;- vegan::vegdist(otu_table(ps_prop), method = &quot;bray&quot;)</span></span>
<span id="cb39-9"><a href="#cb39-9"></a><span class="co"># as.matrix(bray)[1:6,1:6]</span></span></code></pre></div>
<p>Then perform Principal Coordinate Analysis (PCoA) on your favorite distance matrix. In this case, we are using weighted UniFrac.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>do &lt;-<span class="st"> </span><span class="kw">ordinate</span>(ps_prop, <span class="dt">method =</span> <span class="st">&quot;PCoA&quot;</span>, <span class="dt">distance =</span> wUF)</span>
<span id="cb40-2"><a href="#cb40-2"></a></span>
<span id="cb40-3"><a href="#cb40-3"></a>baseplot &lt;-<span class="st"> </span><span class="kw">plot_ordination</span>(ps_prop, do,</span>
<span id="cb40-4"><a href="#cb40-4"></a>                            <span class="dt">type =</span> <span class="st">&quot;samples&quot;</span>,</span>
<span id="cb40-5"><a href="#cb40-5"></a>                            <span class="dt">color =</span> <span class="st">&quot;Treatment&quot;</span>,</span>
<span id="cb40-6"><a href="#cb40-6"></a>                            <span class="dt">shape =</span> <span class="st">&quot;Timepoint&quot;</span>) <span class="op">+</span></span>
<span id="cb40-7"><a href="#cb40-7"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;HCC ordination, all samples</span><span class="ch">\n</span><span class="st">method=PCoA, distance=weighted UniFrac&quot;</span>) <span class="op">+</span></span>
<span id="cb40-8"><a href="#cb40-8"></a><span class="st">  </span><span class="kw">theme_bw</span>()</span>
<span id="cb40-9"><a href="#cb40-9"></a></span>
<span id="cb40-10"><a href="#cb40-10"></a><span class="co"># Single plot with all samples:</span></span>
<span id="cb40-11"><a href="#cb40-11"></a>baseplot <span class="op">+</span></span>
<span id="cb40-12"><a href="#cb40-12"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">2</span>)</span>
<span id="cb40-13"><a href="#cb40-13"></a>  </span>
<span id="cb40-14"><a href="#cb40-14"></a><span class="co"># Idem, but with sample names on the plot: </span></span>
<span id="cb40-15"><a href="#cb40-15"></a>baseplot <span class="op">+</span></span>
<span id="cb40-16"><a href="#cb40-16"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb40-17"><a href="#cb40-17"></a><span class="st">  </span><span class="kw">geom_label</span>(<span class="kw">aes</span>(<span class="dt">label =</span> SampleID), <span class="dt">size =</span> <span class="fl">3.5</span>, <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb40-18"><a href="#cb40-18"></a></span>
<span id="cb40-19"><a href="#cb40-19"></a><span class="co"># Facet by time point:</span></span>
<span id="cb40-20"><a href="#cb40-20"></a>baseplot <span class="op">+</span></span>
<span id="cb40-21"><a href="#cb40-21"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb40-22"><a href="#cb40-22"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;HCC ordination, all samples</span><span class="ch">\n</span><span class="st">method=PCoA, distance=weighted UniFrac&quot;</span>) <span class="op">+</span></span>
<span id="cb40-23"><a href="#cb40-23"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Timepoint, <span class="dv">2</span>)</span></code></pre></div>
<p>We can see some grouping by timepoint. Depending on your experiment and questions, you might want to consider other ordination or clustering approaches.</p>
</div>
<div id="permutational-analysis-of-variance-permanova" class="section level3">
<h3>Permutational analysis of variance (PERMANOVA)</h3>
<p>Finally, we’ll test whether there is a significant ecological level treatment effect. <em>PERMANOVA</em> sounds fancy, but it is just an <em>ANOVA</em> performed using permutations. Permutations are used to determine how data may appear if there is no treatment effect and group differences are due to random chance. Observed data are then compared to the randomized data to calculate a p-value.</p>
<p>Treatments are independent within Timepoints, so we will first subset the data by Timepoint:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a>ps_S1 &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(ps_prop, Timepoint <span class="op">==</span><span class="st"> &quot;S1&quot;</span>)</span>
<span id="cb41-2"><a href="#cb41-2"></a>ps_S3 &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(ps_prop, Timepoint <span class="op">==</span><span class="st"> &quot;S3&quot;</span>)</span>
<span id="cb41-3"><a href="#cb41-3"></a>ps_S4 &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(ps_prop, Timepoint <span class="op">==</span><span class="st"> &quot;S4&quot;</span>)</span></code></pre></div>
<p>Subset the metadata and estimate UniFrac for S3:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>metadata &lt;-<span class="st"> </span><span class="kw">as</span>(<span class="kw">sample_data</span>(ps_S3), <span class="st">&quot;data.frame&quot;</span>)</span>
<span id="cb42-2"><a href="#cb42-2"></a>unifrac_dist &lt;-<span class="st"> </span><span class="kw">UniFrac</span>(ps_S3, <span class="dt">weighted =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Then we’ll use <code>vegan::adonis2</code> to perform the <em>PERMANOVA</em>.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a>permanova &lt;-<span class="st"> </span><span class="kw">adonis2</span>(unifrac_dist <span class="op">~</span><span class="st"> </span>Treatment,</span>
<span id="cb43-2"><a href="#cb43-2"></a>                     <span class="dt">data =</span> metadata,</span>
<span id="cb43-3"><a href="#cb43-3"></a>                     <span class="dt">permutations =</span> <span class="dv">10000</span>)</span>
<span id="cb43-4"><a href="#cb43-4"></a></span>
<span id="cb43-5"><a href="#cb43-5"></a>permanova</span></code></pre></div>
<p>The low p-value suggests that within the S3 timepoint (post-termination), there is a significant treatment effect.</p>
<hr />
</div>
</div>
<div id="differential-taxon-abundance-deseq" class="section level2">
<h2>Differential taxon abundance (DESeq)</h2>
<p>Here, we will use <a href="https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html">DESeq2</a> to identify differentially abundant taxa between time points and treatments. <em>DESeq2</em> was designed to analyze RNAseq datasets, which are similar to OTU/ASV data sets in that both handle large, sparse contingency tables generated from Illumina sequencing data. For more details, see <a href="https://dx.plos.org/10.1371/journal.pcbi.1003531">McMurdie and Holmes 2014</a>.</p>
<p>First, we’ll convert our non-normalized count data to a <em>DESeq</em> object.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a>dds &lt;-<span class="st"> </span><span class="kw">phyloseq_to_deseq2</span>(ps, <span class="op">~</span>Timepoint <span class="op">+</span><span class="st"> </span>Treatment)</span>
<span id="cb44-2"><a href="#cb44-2"></a></span>
<span id="cb44-3"><a href="#cb44-3"></a>dds</span>
<span id="cb44-4"><a href="#cb44-4"></a><span class="kw">colData</span>(dds)</span></code></pre></div>
<p>There are two ways to analyze interaction effects using <em>DESeq2</em>. The first is to fit a multivariate model (e.g. ~A+B+A:B) and explore the model coefficients. The second is to fit a univariate model and explore pairwise contrasts. Here, we will group our two factors (Treatment and Timepoint) and use the latter approach.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a><span class="co"># Prepare the data:</span></span>
<span id="cb45-2"><a href="#cb45-2"></a>dds<span class="op">$</span>Group &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">paste0</span>(dds<span class="op">$</span>Treatment, dds<span class="op">$</span>Timepoint))</span>
<span id="cb45-3"><a href="#cb45-3"></a><span class="kw">design</span>(dds) &lt;-<span class="st"> </span><span class="kw">formula</span>(<span class="op">~</span>Group)</span>
<span id="cb45-4"><a href="#cb45-4"></a>dds &lt;-<span class="st"> </span>dds[, <span class="dv">-50</span>]                    <span class="co"># Remove tech_rep sample</span></span>
<span id="cb45-5"><a href="#cb45-5"></a>dds<span class="op">$</span>Group &lt;-<span class="st"> </span><span class="kw">droplevels</span>(dds<span class="op">$</span>Group)   <span class="co"># Drop tech_rep level from design matrix</span></span>
<span id="cb45-6"><a href="#cb45-6"></a></span>
<span id="cb45-7"><a href="#cb45-7"></a><span class="co"># Perform the differential abundance analysis:</span></span>
<span id="cb45-8"><a href="#cb45-8"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds)</span>
<span id="cb45-9"><a href="#cb45-9"></a></span>
<span id="cb45-10"><a href="#cb45-10"></a><span class="kw">results</span>(dds)</span></code></pre></div>
<p>Let’s explore the contrast, “T1S4 : T4S4” (no cc, seedling stage : late termination, seedling stage):</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a>res &lt;-<span class="st"> </span><span class="kw">results</span>(dds, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">&quot;Group&quot;</span>, <span class="st">&quot;T1S4&quot;</span>, <span class="st">&quot;T4S4&quot;</span>))</span>
<span id="cb46-2"><a href="#cb46-2"></a><span class="kw">summary</span>(res)</span>
<span id="cb46-3"><a href="#cb46-3"></a></span>
<span id="cb46-4"><a href="#cb46-4"></a><span class="co"># Number of taxa with significantly higher abundance in T4S4:</span></span>
<span id="cb46-5"><a href="#cb46-5"></a><span class="kw">sum</span>(res<span class="op">$</span>padj <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.1</span> <span class="op">&amp;</span><span class="st"> </span>res<span class="op">$</span>log2FoldChange <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</span>
<span id="cb46-6"><a href="#cb46-6"></a></span>
<span id="cb46-7"><a href="#cb46-7"></a><span class="co"># Number of taxa with significantly lower abundance in T4S4:</span></span>
<span id="cb46-8"><a href="#cb46-8"></a><span class="kw">sum</span>(res<span class="op">$</span>padj <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.1</span> <span class="op">&amp;</span><span class="st"> </span>res<span class="op">$</span>log2FoldChange <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span>, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>We can sort results by p-value and look at the top ASVs:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a>res.order &lt;-<span class="st"> </span>res[<span class="kw">order</span>(res<span class="op">$</span>pvalue), ]</span>
<span id="cb47-2"><a href="#cb47-2"></a><span class="kw">head</span>(res.order)[<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dv">6</span>)]</span></code></pre></div>
<p>And plot counts for the first in the ordered list ASV:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1"></a><span class="kw">plotCounts</span>(dds, <span class="dt">gene =</span> res.order<span class="op">@</span>rownames[<span class="dv">1</span>], <span class="dt">intgroup =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p>We can also plot counts by sample_ID:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="kw">plotCounts</span>(dds, <span class="dt">gene =</span> <span class="st">&quot;ASV_48&quot;</span>, <span class="dt">intgroup =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p>The plot can also be saved and customized with ggplot:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a>d &lt;-<span class="st"> </span><span class="kw">plotCounts</span>(dds, <span class="dt">gene =</span> <span class="st">&quot;ASV_48&quot;</span>, <span class="dt">intgroup =</span> <span class="st">&quot;Group&quot;</span>, <span class="dt">returnData =</span> <span class="ot">TRUE</span>)</span>
<span id="cb50-2"><a href="#cb50-2"></a></span>
<span id="cb50-3"><a href="#cb50-3"></a><span class="kw">ggplot</span>(d, <span class="kw">aes</span>(<span class="dt">x =</span> Group, <span class="dt">y =</span> count)) <span class="op">+</span></span>
<span id="cb50-4"><a href="#cb50-4"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">position =</span> <span class="kw">position_jitter</span>(<span class="dt">w =</span> <span class="fl">0.1</span>, <span class="dt">h =</span> <span class="dv">0</span>), <span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></span>
<span id="cb50-5"><a href="#cb50-5"></a><span class="st">  </span><span class="kw">scale_y_log10</span>() <span class="op">+</span></span>
<span id="cb50-6"><a href="#cb50-6"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="kw">expression</span>(<span class="kw">paste</span>(<span class="st">&quot;ASV 48, log&quot;</span>[<span class="dv">10</span>], <span class="st">&quot;-scale&quot;</span>))) <span class="op">+</span></span>
<span id="cb50-7"><a href="#cb50-7"></a><span class="st">  </span><span class="kw">theme_bw</span>()</span></code></pre></div>
<p>We can also subset our differentially abundant taxa to keep those with an adjusted p-value &lt; 0.05:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a>sig &lt;-<span class="st"> </span><span class="kw">subset</span>(res.order, padj <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>)</span>
<span id="cb51-2"><a href="#cb51-2"></a><span class="kw">head</span>(sig)</span></code></pre></div>
<p>What taxa are associated with significantly differentially abundant ASVs?</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1"></a>sig_asv &lt;-<span class="st"> </span><span class="kw">rownames</span>(sig)</span>
<span id="cb52-2"><a href="#cb52-2"></a>sig_taxa &lt;-<span class="st"> </span><span class="kw">tax_table</span>(ps)[sig_asv,]</span>
<span id="cb52-3"><a href="#cb52-3"></a><span class="kw">head</span>(sig_taxa)</span>
<span id="cb52-4"><a href="#cb52-4"></a></span>
<span id="cb52-5"><a href="#cb52-5"></a>sig_write &lt;-<span class="st"> </span><span class="kw">cbind</span>(sig, sig_taxa)</span>
<span id="cb52-6"><a href="#cb52-6"></a><span class="kw">write.csv</span>(sig_write, <span class="kw">file.path</span>(outdir, <span class="st">&#39;DATaxa_T1S1_T1S4.csv&#39;</span>))</span></code></pre></div>
<p>Report that we are done!</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r r_code"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a><span class="kw">print</span>(<span class="st">&#39;Done with ASV inference.&#39;</span>)</span></code></pre></div>
<hr />
</div>
<div id="resources" class="section level2">
<h2>Resources</h2>
<ul>
<li><p><a href="https://benjjneb.github.io/decontam/vignettes/decontam_intro.html"><code>decontam</code> tutorial</a></p></li>
<li><p><a href="https://peat-clark.github.io/BIO381/veganTutorial.html"><code>vegan</code> tutorial</a></p></li>
<li><p><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html"><code>DESeq2</code> tutorial</a></p></li>
<li><p><a href="https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/">Intro course on microbiome analysis</a></p></li>
<li><p><a href="https://doi.org/10.1128/AEM.71.12.8228-8235.2005">UniFrac publication</a></p></li>
<li><p><a href="https://doi.org/10.1128/AEM.01996-06">Additional UniFrac publication</a></p></li>
<li><p><strong>Articles on data normalization:</strong></p>
<ul>
<li><a href="https://dx.plos.org/10.1371/journal.pcbi.1003531">Waste not, want not: why rarifying microbiome data is inadmissible</a></li>
<li><a href="http://journal.frontiersin.org/article/10.3389/fmicb.2017.02224/full">Microbiome datasets are compositional: and this is not optional</a></li>
<li><a href="http://doi.wiley.com/10.1111/2041-210X.13115">Methods for normalizing microbiome data: an ecological perspective</a></li>
<li><a href="http://www.nature.com/articles/nmeth.2658">Differential abundance analysis for microbial marker-gene surveys</a></li>
</ul></li>
</ul>
</div>

<div id="rmd-source-code">---
title: "<br>Workflow part III:<br>Analyzing ASV Data"
output:
  html_document:
    theme: cerulean
    highlight: tango
    code_download: true
    toc: true
    toc_float: true
    css: my.css
editor_options: 
  chunk_output_type: console
---

```{r knitr_options, echo=FALSE, purl=FALSE}
root_dir <- "/fs/project/PAS0471/workshops/2020-12_micro/$USER"
#knitr::opts_knit$set(root.dir = root_dir) # UNCOMMENT AT OSC

knitr::opts_chunk$set(results='hide',
                      eval = FALSE, # CHANGE TO TRUE TO RUN UPON COMPILING
                      class.source="r_code",
                      class.output="r_output",
                      class.warning="r_warning",
                      class.message="r_warning",
                      class.error="r_error")
```

<br>

-----

## Goals

Analyze microbiome experimental data as a phyloseq object - explore ecological metrics and identify differentially abundant taxa.

-----

## Before We Get Started

### Notes

- This is a slightly modified version of a document created by [Matthew Willman](mailto:willman.18@osu.edu).

- To convert an `Rmd` (R Markdown) file to an R script,
  type the following in an R console:
  `knitr::purl(input="<filename>.Rmd")`.
  (You can download this `Rmd` file by clicking the `Code` button
  in the top-right of this page -- but we will open it at OSC.)

### Open this file in RStudio Server at OSC

- This assumes you still have an active RStudio Server job at OSC;
   if not see the
   [instructions on the previous page](07-ASV-inference.html#start-an-rstudio-server-job-at-osc) to start one.

- In the Files pane, in the `markdown` directory,
  open the file `08-postASV-analysis.Rmd`. That's this file!

 
-----

## Getting Started

### Install and Load Packages

Add our custom library again if you restarted your R session:

```{r libload}
.libPaths(new = '/fs/project/PAS0471/.R/4.0/')
```

And load the necessary packages:

```{r packages, message=FALSE, warning=FALSE}
packages <- c('tidyverse', 'vegan', 'phyloseq', 'decontam', 'ape',
              'DESeq2', 'microbiome', 'metagenomeSeq', 'remotes',
              'ampvis2', 'breakaway')
pacman::p_load(char = packages)

# If you wanted to install and load these packages yourself,
# first make sure you have the pacman package installed:
## install.packages('pacman')
# Then, the code above would work except for the last two packages,
# which are available from Github only. To install these, you would run:
## remotes::install_github("MadsAlbertsen/ampvis2")
## remotes::install_github("adw96/breakaway")
```

### Set Directories

```{r}
# Dir with input files:
indir <- 'analysis/ASV_inference/'

# Dir for output:
outdir <- 'analysis/postASV_analysis/'
```

### Load the Data

This session starts with a *phyloseq* object similar to the one generated using
*DADA2* in the [previous session](06-reads-to-ASV.Rmd).

```{r data}
ps_raw <- readRDS(file.path(indir, 'ps_16S_V4_withtree.rds'))

# This would load the object from the previous session, which doesn't contain a tree:
# ps_raw <- readRDS(file.path(indir, 'ps_V4.rds'))
```

Our *phyloseq* object is made up of the following components (slots):

- `otu_table`
- `sample_data`
- `tax_table`
- `phy_tree`

Let's have a look at the object:

```{r ps}
ps_raw

# And the assigned taxon names:
taxa_names(ps_raw)[1:3]
```

As we can see, the taxon names are currently the associated sequences.
We can create a new *phyloseq* component to store these sequences,
then rename the ASVs so something shorter ("ASV_1", "ASV_2", etc.).

```{r}
dna <- Biostrings::DNAStringSet(taxa_names(ps_raw))
names(dna) <- taxa_names(ps_raw)

# Merge sequence object into the phyloseq object:
ps_raw <- merge_phyloseq(ps_raw, dna)
ps_raw

# Rename ASVs:
taxa_names(ps_raw) <- paste("ASV", 1:ntaxa(ps_raw), sep = "_")
taxa_names(ps_raw)[1:3]
```

-----

## Filter taxa

### Identify and Remove Contaminants

It is possible to introduce contaminating microbes during sample preparation.
Before analyzing the data, we will identify and remove probable contaminants using
the [decontam package](https://benjjneb.github.io/decontam/vignettes/decontam_intro.html).   

In this case, we will define a conataminant as an ASV whose abundance correlates
with DNA concentration (post-PCR).
Our assumption here is that each soil taxon's abundance should be independant of
DNA concentration. However, if we were to spike each sample with a contaminant,
that contaminant would show a greater abundance in samples with lower DNA concentrations.
To do this, we need DNA concentration measured after PCR and before pooling of the samples.

These data are in column 7 of the sample data component:

```{r DNA_concentration}
head(sample_data(ps_raw))
```

MCIC measured our DNA concentration by comparing band intensities to a single
reference band. MCIC's DNA concentration measurements range from 0 to 1 (the reference).
Zero in this case is coded as `NA`.
The *decontam* package can't handle NA's or 0's,
so we need to change them to a small value (e.g. 0.01).

```{r}
# First, we identify the NAs in the DNA measurement column:
NAs <- is.na(sample_data(ps_raw)$PCR_prod_V4.V5)

# Then, we replace NAs with 0.01:
sample_data(ps_raw)$PCR_prod_V4.V5[NAs] <- 0.01
```

We will use `decontam::isContaminant(method = "frequency")` to test each taxon
against the null hypothesis: abundance is not associated with DNA concentration.
Because the DNA concentration measurements are not very accurate,
and we want to make sure to remove possible contaminants,
we use a relaxed p-value (0.2) to reject the null hypothesis:

```{r, cache=TRUE}
contam_df_freq <- isContaminant(ps_raw,
                                method = "frequency",
                                conc = "PCR_prod_V4.V5",
                                threshold = 0.2)
head(contam_df_freq)
```

As we can see, the top 3 ASVs are identified as probable contaminants.
How many contaminants were identified (`TRUE` in the `contaminant` column)
in total?

```{r}
table(contam_df_freq$contaminant)
```

What are the most abundant contaminant taxa?

```{r}
# Get abundance ranks for contaminant ASVs:
head(which(contam_df_freq$contaminant))

# Check which taxa are contaminants:
ps_contam <- prune_taxa(contam_df_freq$contaminant, ps_raw)
head(tax_table(ps_contam))
```

Let's take a look at our tested associations for several taxa:

```{r, warning=FALSE}
plot_frequency(ps_raw,
               taxa_names(ps_raw)[c(1, 2, 3, 4, 5, 11)],
               conc = "PCR_prod_V4.V5") +
  xlab("DNA Concentration (relative to single marker)")
```
   
As we can see, abundance of ASVs 1, 2, 3, and 11 are particularly high in samples
with low DNA concentration.

Which samples have highest abundances of the contaminants?

```{r}
asv1 <- otu_table(ps_raw)[, 1]
head(asv1[order(asv1, decreasing = TRUE)])

asv2 <- otu_table(ps_raw)[, 2]
head(asv2[order(asv2, decreasing = TRUE)])
```

How do these contaminants overlap with ASVs present in the negative control?

```{r}
otu_table(ps_raw)[70:75, 1:11]
```

ASVs have high read counts in the extraction negative control.
Thus, we can be fairly certain these are contaminants and not representative
of our experimental samples.   
   
Remove the contaminants that we identified:

```{r prune_decontam}
ps_noncontam <- prune_taxa(!contam_df_freq$contaminant, ps_raw)

ps_noncontam
```

What proportion of our count data were removed as contaminants?

```{r count_decontam1}
pre <- sum(sample_sums(ps_raw))
post <- sum(sample_sums(ps_noncontam))

(pre-post) / pre
```

### Remove non-bacterial, non-archaeal ASVs

The V4 region of 16S rRNA is conserved within certain bacteria, archaea,
chloroplasts, mitochondria, and eukaryotes.
Since we are only interested in bacteria and archaea here,
we'll need to remove other taxa from our object.

```{r remove_nonbac}
# Create ps subsets for chloroplast, mitochondria, and eukaryotes:
chlr <- subset_taxa(ps_noncontam, Order == "Chloroplast")
mit <- subset_taxa(ps_noncontam, Family == "Mitochondria")
euk <- subset_taxa(ps_noncontam, Kingdom == "Eukaryota")

# Get all taxon IDs that we want to remove:
bad_taxa <- c(taxa_names(chlr), taxa_names(mit), taxa_names(euk))

# Get all taxon IDs that we want to keep:
all_taxa <- taxa_names(ps_noncontam)
good_taxa <- all_taxa[!(all_taxa %in% bad_taxa)]

# Subset phyloseq object:
ps_bac_arc <- prune_taxa(good_taxa, ps_noncontam)
```

What proportion of ASVs were kept?

```{r kept_bac}
pre <- sum(sample_sums(ps_noncontam))
post <- sum(sample_sums(ps_bac_arc))
post / pre
```
   
Finally, we'll also remove non-experimental samples (e.g. negative controls).
Note, though, that non-experimental samples are important for evaluating the
quality of your reagents, consistency across sequencing runs, and contamination.

```{r subset_HCC}
ps <- subset_samples(ps_bac_arc, Experiment == "HCC")

ps
```

-----

## Filter samples

After having filtered out unwanted taxa,
we now need to filter out uninformative samples --
those with low total taxon counts.

First, how many counts do we have for each sample?

```{r sums}
sums <- sample_sums(ps)
sums[order(sums)]

# Lowest total count:
min(sums)

# Sample with lowest counts:
names(sums[order(sums)][1])
```

To remove uninformative samples, we will only keep those with over 1,000 counts.
We will also remove a technical replicate,
though note that technical replicates are important for quality control of runs
and for understanding the reproducibility of the methodology. 

```{r subset_1k}
# Remove samples with low counts:
ps <- subset_samples(ps, sample_sums(ps) > 1000)

# Remove the technical replicate:
ps <- subset_samples(ps, SampleID != "501S4")
```

Let's take a look at our current `phyloseq` object:

```{r look_ps}
sample_data(ps)[1:10]

head(t(otu_table(ps)[1:10]))

head(tax_table(ps))
```

-----

## Normalization

Normalization is currently a much-discussed issue of microbiome studies.
Differences in read depth between samples often need to be corrected before analysis.
Several normalization methods have been proposed, and no single method is perfect.
It may be that the most appropriate method depends on the analysis.

In this tutorial, we will use:
  
  - Proportion-normalized data to estimate ecological metrics.
  
  - Stabilizing transformation normalized data to identify differentially abundant taxa.

However, for your reference, we will also present code for two other normalization
methods: rarefaction and cumulative sum scaling.

For more details, read
[McMurdie and Holmes 2014](https://dx.plos.org/10.1371/journal.pcbi.1003531)
and [McKnight et al. 2019](http://doi.wiley.com/10.1111/2041-210X.13115).

-----

### Proportion Normalization

Proportion normalization involves dividing each OTU count by the total sum for each sample.
The resulting count data will add up to 1 (100%) for each sample.

The `microbiome::transform` function can be used to easily normalize count data as
proportions in a *phyloseq* object:

```{r norm_proportions}
# Proportion normalization:
ps_prop <- transform(ps, "compositional")

# Have a look at the resulting OTU table:
otu_table(ps_prop)[1:5, 1:5]

# Check what the sums for each sample are now:
head(sample_sums(ps_prop))
```

-----

### Variance stabilizing transformation

This transformation method, available in [DESeq2](https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html),
normalizes data with respect to library size as well as dispersion.   
Type `?vst` or read [Anders and Huber 2010](http://dx.doi.org/10.1186/gb-2010-11-10-r106)
for more information.

```{r norm_vst, message=FALSE, warning=FALSE}
# First we convert the phyloseq object to a DESeq object:
dds <- phyloseq_to_deseq2(ps, ~1)

# Then we estimate the size factors:
dds <- estimateSizeFactors(dds)

# Now can do the transformation:
dds_vst <- vst(dds, blind = TRUE)

# Have a look at the resulting counts:
vst_counts <- assay(dds_vst)
t(vst_counts)[1:5, 1:5]
```

-----

### Rarefaction

Rarefaction can be used to subset data such that the library depth is the same
for each sample.
Because sampling of the data is random,
rarefaction can account for an effect of total read count on taxa richness.
However, rarefaction is no longer considered to be a good way to normalize
amplicon sequencing data
(see [McMurdie & Holmes 2014](https://dx.plos.org/10.1371/journal.pcbi.1003531)).

```{r, rarefy, message=FALSE, warning=FALSE}
ps_rarefied <- rarefy_even_depth(ps,
                                 rngseed = 1,
                                 sample.size = min(sample_sums(ps)),
                                 replace = FALSE)

# Have a look at the result:
otu_table(ps_rarefied)[1:5, 1:5]

head(sample_sums(ps_rarefied))
```

Compare these sample sums to the non-rarified data sums:

```{r, non-rar_sums}
head(sample_sums(ps))
```

-----

### Cumulative Sum Scaling

The *metagenomeSeq* Cumulative Sum Scaling (CSS) normalization is another option
developed for microbiome data.
For more information, read
[Paulson et al. 2013](http://www.nature.com/articles/nmeth.2658).

```{r norm_css, message=FALSE}
# Convert the phyloseq object to a metagenomeseq object:
mgs_css <- phyloseq_to_metagenomeSeq(ps)

# Perform the Cumulative Sum Scaling:
mgs_css <- cumNorm(mgs_css)

# Extract the counts and add them to a separate phyloseq object:
css_counts <- MRcounts(mgs_css, norm = TRUE)
ps_css <- ps
otu_table(ps_css) <- otu_table(t(css_counts), taxa_are_rows = FALSE)
```

Now lets compare the original data to the CSS normalized data:

```{r ccs_compare}
otu_table(ps)[1:5, 1:5]
head(sample_sums(ps))

otu_table(ps_css)[1:5, 1:5]
head(sample_sums(ps_css))
```

-----

## Plot abundances

In this and following the examples we will use the **proportions** data.

```{r plot_abundance}
# Subset to the T1 treatment
T1 <- subset_samples(ps_prop, Treatment == "T1")

# Agglomerate taxa to phylum level:
T1_phylum <- tax_glom(T1, taxrank = "Phylum", NArm = FALSE)

# Remove uncommon phyla:
T1_phylum <- subset_taxa(T1_phylum, taxa_sums(T1_phylum) > 0.1)

# Plot:
plot_bar(T1_phylum, fill = "Phylum") +
   facet_wrap(~Timepoint, scales = "free_x", nrow = 1)
```
  
-----

## Alpha diversity

Alpha diversity measures taxonomic diversity within a single population.
Measures of alpha diversity include taxonomic richness (i.e. number of taxa),
and indices which combine taxonomic richness with some measure of evenness.   
   
*Note:* Many of these methods are greatly influenced by singleton data
(i.e. taxa represented by a single count), meaning they may be unreliable if your
data excludes singletons.
DADA2 does not output singletons if it is run individually on each sample.
Thus, if using dada2 to infer ASVs, it is advisable to use `pool = TRUE` when
running the dada algorithm.
This will allow calling per-sample singletons, but not per-study singletons.   
   
### Taxonomic richness: Rarefaction plot

Our goal in defining richness is to determine the number of unique taxa within
each population (in our case, each plot).
This is difficult, as our samples contain only a portion of the total richness
within our plots.
As sequence depth increases for each sample, so does the number of taxa.
This can be illustrated with a rarefaction plot.   
   
To make a rarefaction plot, we draw random samples from our data and count the
number of unique ASVs as samples are drawn.
The resulting rarefaction curve is expected to rise quickly then plateu as the
most abundant taxa are represented.
We can make a quick rarefaction curve plot directly from our `phyloseq` object of
all samples using the
[`vegan` package](https://peat-clark.github.io/BIO381/veganTutorial.html):

```{r rarecurve}
rarecurve(otu_table(ps), step = 500, xlab = "Sample Size", ylab = "Taxa")
```
   
We can also split rarefaction curves by group using the
[`ampvis2` package](https://madsalbertsen.github.io/ampvis2/articles/ampvis2.html).

```{r rarefaction1}
metadata <- data.frame(sample_data(ps), check.names = FALSE)

asv_table <- data.frame(t(otu_table(ps)),
                       tax_table(ps),
                       check.names = FALSE)
```

`ampvis2` requires a `Species` column, which must be added to our data.
If your table already has a `Species` column, skip this step:

```{r add_Species}
asv_table$Species <- NA
```

Load an `ampvis2` object and calculate rarefaction:

```{r rarecurve_ampvis2, cache=TRUE, warning=FALSE}
ps3 <- amp_load(asv_table, metadata)

rar_plot <- amp_rarecurve(ps3,
                          stepsize = 100,
                          facet_by = "Timepoint",
                          color_by = "Treatment") +
   ylab("Number of observed ASVs")

rar_plot
```
   
As we can see, as read depth increases for each sample, so does taxonomic richness.
In all samples, increased sampling would result in increased richness.
Thus, our observed sample richness is lower than the true population richness for
every plot.

Further, sequence depth and total richness appear to be correlated across samples:

```{r corr, warning=FALSE}
dp_by_asv <- data.frame(ct_sums = sample_sums(ps),
                        asvs = rowSums(otu_table(ps) != 0))

ggplot(data = dp_by_asv) +
  geom_point(aes(x =  ct_sums, y = asvs)) +
  labs(x = "Sequence depth", y= "ASV count") +
  theme_minimal()
```

We can test whether this correlation is significant:

```{r}
cor <- cor.test(dp_by_asv$ct_sums, dp_by_asv$asvs, method = "spearman")
cor
```

Here we observe a significant correlation between read count and observed
taxonomic richness.   
   
Thus, a sample from a population with high actual diversity, but low read depth,
could have low observed richness.
Fortunately, there are methods to estimate the number of unobserved taxa.
For an overview of these methods, read
[Bunge et al. 2014](http://www.annualreviews.org/doi/10.1146/annurev-statistics-022513-115654).
   
### Frequency counts

To illustrate the difficulty of estimating population richness and importance of
singletons to alpha diversity metrics,
we'll look at the frequency count distribution for our first sample: 101-S1.

We'll count the number of singletons, doubletons, etc.,
then plot frequency as a function of count.

```{r freq_ct, message=FALSE}
# Get the frequency count distribution for the 1st sample:
frequencytablelist <- build_frequency_count_tables(otu_table(ps))
freq_ct <- frequencytablelist[[1]]
colnames(freq_ct) <- c("Frequency", "Count")

# Plot:
ggplot(freq_ct, aes(Frequency, Count)) + 
  geom_point() + 
  scale_x_continuous(breaks = seq(0, 600, by = 50)) + 
  ggtitle("101-S1") +
  theme_minimal()
```
   
We can see there are ~3,200 taxa represented as singletons,
and ~1,600 taxa as doubletons. 
This is a high proportion of the total number taxa are observed in this sample:

```{r}
# Total number of taxa:
sum(freq_ct$Count)

# Proportion of singletons
freq_ct$Count[1] / sum(freq_ct$Count)
```

How many more taxa might we observe if we increased our sampling?   

### Diversity Indices

We can use several indices to explore how evenly species are distributed within
each sample. Note that 'Observed' is the sample taxonomic richness.
The other three indices combine richness and abundance data for each taxon.

```{r alpha_div}
plot_richness(ps,
              x = "Timepoint",
              shape = "Treatment",
              color = "Treatment", 
              measures = c("Observed", "Shannon", "Simpson", "InvSimpson"))
```
   
Notice that several T2:S3 samples have low measures,
suggesting they have low diversity (i.e. abundances are dominated by a few taxa)
compared to the other samples.
 
-----

## Beta diversity

Beta diversity measures differences in microbial compositions between populations.
If available (use ps_16S_V4_withtree.rds), we can use a phylogenetic tree and otu
table to estimate the [UniFrac](https://doi.org/10.1128/AEM.71.12.8228-8235.2005)
for each sample pair.

The result is a distance matrix with height and width equal to the number of samples.
UniFrac may be [weighted or unweighted](https://doi.org/10.1128/AEM.01996-06).
Selection of weighted or unweighted will depend on the question you want to answer.
Unweighted UniFrac uses only presence or absence of a taxon from the otu table
(i.e. 1000 will be treated the same as 1).
It is appropriate if you want to test qualitative taxa differences between samples.
Weighted UniFrac uses presence as well as quantity data (i.e. 1 < 1000).

We are interested to know what changes to taxon abundance may be driven by our treatments, so we will use weighted UniFrac.    

### Ordination

First we'll make a weighted UniFrac distance matrix.
For UniFrac, you require a tree as part of your *phyloseq* object.

```{r wUF, warning=FALSE, cache=FALSE}
set.seed(100)
wUF <- phyloseq::distance(ps_prop, method = "wunifrac")

as.matrix(wUF)[1:6, 1:6]

# Note: If your *phyloseq* object lacks a phylogenetic tree,
# you could use a dissimilarity index that does not require one, such as Bray-Curtis:
# bray <- vegan::vegdist(otu_table(ps_prop), method = "bray")
# as.matrix(bray)[1:6,1:6]
```

Then perform Principal Coordinate Analysis (PCoA) on your favorite distance matrix.
In this case, we are using weighted UniFrac.

```{r ordinate, warning=FALSE, cache=FALSE}
do <- ordinate(ps_prop, method = "PCoA", distance = wUF)

baseplot <- plot_ordination(ps_prop, do,
                            type = "samples",
                            color = "Treatment",
                            shape = "Timepoint") +
  ggtitle("HCC ordination, all samples\nmethod=PCoA, distance=weighted UniFrac") +
  theme_bw()

# Single plot with all samples:
baseplot +
  geom_point(size = 2)
  
# Idem, but with sample names on the plot: 
baseplot +
  geom_point(size = 1) + 
  geom_label(aes(label = SampleID), size = 3.5, color = "black")

# Facet by time point:
baseplot +
  geom_point(size = 2) + 
  ggtitle("HCC ordination, all samples\nmethod=PCoA, distance=weighted UniFrac") +
  facet_wrap(~Timepoint, 2)
```
   
We can see some grouping by timepoint.
Depending on your experiment and questions,
you might want to consider other ordination or clustering approaches.

### Permutational analysis of variance (PERMANOVA)

Finally, we'll test whether there is a significant ecological level treatment effect.
*PERMANOVA* sounds fancy, but it is just an *ANOVA* performed using permutations.
Permutations are used to determine how data may appear if there is no treatment
effect and group differences are due to random chance. Observed data are then
compared to the randomized data to calculate a p-value.   
   
Treatments are independent within Timepoints,
so we will first subset the data by Timepoint:

```{r subset_Timepoints, warning=FALSE}
ps_S1 <- subset_samples(ps_prop, Timepoint == "S1")
ps_S3 <- subset_samples(ps_prop, Timepoint == "S3")
ps_S4 <- subset_samples(ps_prop, Timepoint == "S4")
```

Subset the metadata and estimate UniFrac for S3:

```{r set_perm, warning=FALSE}
metadata <- as(sample_data(ps_S3), "data.frame")
unifrac_dist <- UniFrac(ps_S3, weighted = TRUE)
```

Then we'll use `vegan::adonis2` to perform the *PERMANOVA*.

```{r PERMANOVA}
permanova <- adonis2(unifrac_dist ~ Treatment,
                     data = metadata,
                     permutations = 10000)

permanova
```

The low p-value suggests that within the S3 timepoint (post-termination),
there is a significant treatment effect.   
 
-----

## Differential taxon abundance (DESeq)

Here, we will use
[DESeq2](https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html)
to identify differentially abundant taxa between time points and treatments.
*DESeq2* was designed to analyze RNAseq datasets, which are similar to OTU/ASV data 
sets in that both handle large, sparse contingency tables generated from Illumina
sequencing data.
For more details, see
[McMurdie and Holmes 2014](https://dx.plos.org/10.1371/journal.pcbi.1003531).  
  
First, we'll convert our non-normalized count data to a *DESeq* object.

```{r ps2dss, message=FALSE, warning=FALSE}
dds <- phyloseq_to_deseq2(ps, ~Timepoint + Treatment)

dds
colData(dds)
```

There are two ways to analyze interaction effects using *DESeq2*.
The first is to fit a multivariate model (e.g. ~A+B+A:B) and explore the model
coefficients.
The second is to fit a univariate model and explore pairwise contrasts.
Here, we will group our two factors (Treatment and Timepoint) and use the latter
approach.

```{r deg, message=FALSE, cache=FALSE}
# Prepare the data:
dds$Group <- factor(paste0(dds$Treatment, dds$Timepoint))
design(dds) <- formula(~Group)
dds <- dds[, -50]                    # Remove tech_rep sample
dds$Group <- droplevels(dds$Group)   # Drop tech_rep level from design matrix

# Perform the differential abundance analysis:
dds <- DESeq(dds)

results(dds)
```
    
Let's explore the contrast, "T1S4 : T4S4"
(no cc, seedling stage : late termination, seedling stage):

```{r}
res <- results(dds, contrast = c("Group", "T1S4", "T4S4"))
summary(res)

# Number of taxa with significantly higher abundance in T4S4:
sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm = TRUE)

# Number of taxa with significantly lower abundance in T4S4:
sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm = TRUE)
```

We can sort results by p-value and look at the top ASVs:

```{r}
res.order <- res[order(res$pvalue), ]
head(res.order)[c(1:2, 6)]
```

And plot counts for the first in the ordered list ASV:

```{r plot_counts}
plotCounts(dds, gene = res.order@rownames[1], intgroup = "Group")
```
   
We can also plot counts by sample_ID:

```{r plot_counts2}
plotCounts(dds, gene = "ASV_48", intgroup = "Group")
```

The plot can also be saved and customized with ggplot:

```{r}
d <- plotCounts(dds, gene = "ASV_48", intgroup = "Group", returnData = TRUE)

ggplot(d, aes(x = Group, y = count)) +
  geom_point(position = position_jitter(w = 0.1, h = 0), size = 2) +
  scale_y_log10() +
  ggtitle(expression(paste("ASV 48, log"[10], "-scale"))) +
  theme_bw()
```
   
We can also subset our differentially abundant taxa to keep those with an
adjusted p-value < 0.05:

```{r filter_05}
sig <- subset(res.order, padj <= 0.05)
head(sig)
```

What taxa are associated with significantly differentially abundant ASVs?

```{r sig_taxa}
sig_asv <- rownames(sig)
sig_taxa <- tax_table(ps)[sig_asv,]
head(sig_taxa)

sig_write <- cbind(sig, sig_taxa)
write.csv(sig_write, file.path(outdir, 'DATaxa_T1S1_T1S4.csv'))
```

Report that we are done!

```{r report_done}
print('Done with ASV inference.')
```

-----

## Resources

- [`decontam` tutorial](https://benjjneb.github.io/decontam/vignettes/decontam_intro.html)

- [`vegan` tutorial](https://peat-clark.github.io/BIO381/veganTutorial.html)

- [`DESeq2` tutorial](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

- [Intro course on microbiome analysis](https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/)

- [UniFrac publication](https://doi.org/10.1128/AEM.71.12.8228-8235.2005)
- [Additional UniFrac publication](https://doi.org/10.1128/AEM.01996-06)

- **Articles on data normalization:**   
   - [Waste not, want not: why rarifying microbiome data is inadmissible](https://dx.plos.org/10.1371/journal.pcbi.1003531)
   - [Microbiome datasets are compositional: and this is not optional](http://journal.frontiersin.org/article/10.3389/fmicb.2017.02224/full)
   - [Methods for normalizing microbiome data: an ecological perspective](http://doi.wiley.com/10.1111/2041-210X.13115)
   - [Differential abundance analysis for microbial marker-gene surveys](http://www.nature.com/articles/nmeth.2658)
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("08-postASV-analysis.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
